---
title: "LungCancer10X_all"
author: "Yiming Zhang"
date: "2019/3/2"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
root.dir = "/mnt/raid62/Lung_cancer_10x/"
knitr::opts_knit$set(root.dir = root.dir)
# setwd(root.dir)


cols = 3
size = 10
dpi = 300
filenames = list()


custom_colors <- c(
    "#1B62A3",
    "#9CB8DD",
    "#EE6719",
    "#F8A966",
    "#278F36",
    "#8CC873",
    "#C70F1F",
    "#EF7E82",
    "#7D509A",
    "#B59BC7",
    "#76423A",
    "#B48880",
    "#CB60A1",
    "#F1A3C5",
    "#7E7F83",
    "#BDC3C4",
    "#ABB126",
    "#D1D57A",
    "#1DAFC3",
    "#8BCFDA"
)
```

```{r}
load_packages <- function() {
    library(Matrix)
    library(Seurat)
    library(here)
    library(data.table)
    library(ggpubr)
    library(openxlsx)
    library(harmony)
    library(dplyr)
}

suppressPackageStartupMessages(load_packages())
```

## Quality control

Filter cells by
1. the gene expresssion of cells >= 1000
2. expressed genes of cells >= 500
3. the sum of gene expression of each cell >= 1000
4. the cells number of each sample >= 200

### read raw counts
```{r cars}
raw_counts <- here("00_data_ingest", "00_raw_data", "raw_counts_clean.csv.gz")

data <- fread(paste("zcat", raw_counts))

# https://www.kaggle.com/cartographic/data-table-to-sparsematrix
to_sparse <- function(d_table){

  i_list <- lapply(d_table, function(x) which(x != 0))
  counts <- unlist(lapply(i_list, length), use.names = F)

  sparseMatrix(
    i = unlist(i_list, use.names = F),
    j = rep(1:ncol(d_table), counts),
    x = unlist(lapply(d_table, function(x) x[x != 0]), use.names = F),
    dims = dim(d_table),
    dimnames = list(NULL, names(d_table)))
}

spareData <- to_sparse(data[ ,2:ncol(data)])
rownames(spareData) <- data$V1

data <- spareData
rm(spareData)
```

## Basic quality validation
```{r}

filenames[[1]] <- here("01_figures", "01_boxplot_gene_expression_across_all_cells.png")

temp <- as.data.frame(rowSums(data))
colnames(temp) <- c("row")
temp$x = "gene"

p <- ggplot(temp, aes(x = x, y=log10(row))) + geom_violin() + geom_boxplot(width=0.1) + labs(x = "", y = "log10(Expression of gene across all cells)")

ggsave(filenames[[1]], plot = p, width = 6, height = 6, dpi = dpi, units = "in")

knitr::include_graphics(filenames[[1]])
```


```{r include=TRUE}
filenames[[2]] <- here("01_figures", "02_boxplot_gene_expression_of_each_cell.png")

temp <- as.data.frame(colSums(data))
colnames(temp) <- c("y")
temp$x = "cells"
p <- ggplot(temp, aes(x = x, y=log10(y))) + geom_violin() + geom_boxplot(width=0.1) + labs(x = "", y = "log10(Expression of each cells)")

ggsave(filenames[[2]], plot = p, width = 6, height = 6, dpi = dpi, units = "in")

knitr::include_graphics(filenames[[2]])
```

```{r fig.height=6, fig.width=12}
filenames[[3]] <- here("01_figures", "03_barplot_num_of_cells_each_sample.png")

temp <- as.data.frame(table(sapply(colnames(data), function(x) {strsplit(x, "_")[[1]][1]})))

p <- ggplot(temp, aes(x=Var1, y=log10(Freq))) + geom_bar(stat = "identity") + theme(axis.text.x = element_text(angle = 90)) + labs(x="Samples", y = "log10(Num. of cells)")

ggsave(filenames[[3]], plot = p, width = 6, height = 6, dpi = dpi, units = "in")

knitr::include_graphics(filenames[[3]])
```

```{r}
# saveRDS(data, here("00_data_ingest", "04_rds_generated", "raw_counts_all.rds"))
```

## Prepare meta info

```{r}
# meta <- read.csv(here("00_data_ingest", "03_annotation_csv", "sample_info_v20190301.csv"), row.names = 1)
# meta[is.na(meta$Batch), "Batch"] <- 3
# meta <- meta[, 1:10]
# 
# 
# sample_with_cell <- as.data.frame(colnames(data))
# colnames(sample_with_cell) <- c("Cell")
# sample_with_cell$SampleID <- sapply(sample_with_cell$Cell, function(x){strsplit(as.character(x), "_")[[1]][1]})
# 
# 
# meta <- merge(meta, sample_with_cell, by = "SampleID")
# 
# saveRDS(meta, here("00_data_ingest", "04_rds_generated", "meta_all.rds"))
```


## Create Seurat obj
```{r}
# obj <- CreateSeuratObject(
#     raw.data = data,
#     project = "LungCancer10X",
#     meta.data = meta
# )
# 
# # Due to the CreateSeuratObject always convert the passed meta into NA
# # set it again
# rownames(meta) <- meta$Cell
# for (i in colnames(meta)) {
#     if (i %in% colnames(obj@meta.data)) {
#         obj@meta.data[, i] <- meta[rownames(obj@meta.data), i]
#     }
# }
```

```{r}
perform_scale <- function(obj, scale.factor=10000, hv.genes.num = 1000, num.cores = 12) {
    obj <- NormalizeData(
        object = obj,
        normalization.method = "LogNormalize",
        scale.factor = scale.factor
    )
    
    obj <- FindVariableGenes(
        object = obj,
        mean.function = ExpMean,
        dispersion.function = LogVMR,
        do.plot = FALSE
    )
    
    hv.genes <- head(rownames(obj@hvg.info), hv.genes.num)
    
    mito.genes <- grep(
        pattern = "^MT-",
        x = rownames(x = obj@raw.data),
        value = TRUE
    )
    
    percent.mito <- Matrix::colSums(obj@raw.data[mito.genes, ])/Matrix::colSums(obj@raw.data)
    
    obj <- AddMetaData(
        object = obj,
        metadata = percent.mito,
        col.name = "percent.mito"
    )
    
    
    obj <- ScaleData(
        object = obj,
        genes.use = hv.genes,
        display.progress = TRUE,
        vars.to.regress = c("percent.mito", "nUMI"),
        do.par = TRUE,
        num.cores = num.cores
    )
    
    obj <- RunPCA(
        object = obj,
        pc.genes = hv.genes,
        pcs.compute = 100,
        do.print = TRUE,
        pcs.print = 1:5,
        genes.print = 1:5
    )
    
    return(obj)
}

# obj <- perform_scale(obj)
```

check the scale
```{r fig.height=6, fig.width=12, include=TRUE}
obj <- readRDS(here("00_data_ingest", "04_rds_generated", "seurat_obj_pcs75_res0.6.rds"))


filenames[[4]] <- here("01_figures", "04_density_boxplot_each_sample.png")


cs <- colSums(obj@raw.data)
cs <- as.data.frame(cs)
cs$sample <- sapply(rownames(cs), function(x) {return(strsplit(x, "_")[[1]][1])})

p1 <- ggplot(
    data = cs,
    aes(x=log10(cs), color = sample)
    ) + geom_density() + theme(legend.position = "none")

cs_scale <- colSums(obj@scale.data)
cs_scale <- as.data.frame(cs_scale)
cs_scale$cs_scale <- cs_scale$cs_scale + abs(min(cs_scale$cs_scale)) + 1
cs_scale$sample <- sapply(rownames(cs_scale), function(x) {return(strsplit(x, "_")[[1]][1])})

p2 <- ggplot(
    data = cs_scale,
    aes(x=log10(cs_scale), color = sample)
    ) + geom_density() + theme(legend.position = "none") + xlim(2.8, NA)


p3 <- ggplot(
    cs,
    aes(x=sample, y=log10(cs), fill=sample)
    ) + geom_violin(trim=TRUE) + theme(
        legend.position = "none",
        axis.text.x = element_text(angle = 45, vjust = 0.5)
        )

p4 <- ggplot(
    cs_scale,
    aes(x=sample, y=log10(cs_scale), fill=sample)
    ) + geom_violin(trim=TRUE) + theme(
        legend.position = "none",
        axis.text.x = element_text(angle = 45, vjust = 0.5)
        )

png(filenames[[4]], width = 20, height = 12, res = dpi, units = "in")
grid.arrange(p1, p2, p3, p4, ncol=1)
dev.off()

```
```{r}
knitr::include_graphics(filenames[[4]])
```

```{r fig.height=6, fig.width=24}

filenames[[5]] <- here("01_figures", "05_vlnplot.png")

p <- VlnPlot(
    object = obj,
    features.plot = c("nGene", "nUMI", "percent.mito"),
    nCol = 3,
    point.size.use=0,
    do.return = TRUE,
    do.sort = TRUE,
    y.log = TRUE
)

ggsave(filename = filenames[[5]], plot = p, dpi = dpi, width = 24, height = 6)

knitr::include_graphics(filenames[[5]])
```

```{r fig.height=6, fig.width=12}
filenames[[6]] <- here("01_figures", "06_geneplot.png")

png(filenames[[6]], res = dpi, width = 12, height = 6, units = "in")
par(mfrow = c(1, 2))
GenePlot(object = obj, gene1 = "nUMI", gene2 = "percent.mito")
GenePlot(object = obj, gene1 = "nUMI", gene2 = "nGene")
dev.off()

knitr::include_graphics(filenames[[6]])
```


```{r}
# obj <- RunPCA(
#     object = obj, 
#     pc.genes = hv.genes, 
#     pcs.compute = 100, 
#     do.print = TRUE, 
#     pcs.print = 1:5, 
#     genes.print = 1:5
# )
```

```{r echo=TRUE}
filenames[[7]] <- here("01_figures", "07_vizpca.png")

png(filenames[[7]], res = dpi, width = 12, height = 6, units = "in")
VizPCA(object = obj, pcs.use = 1:5, nCol = 5)
dev.off()

knitr::include_graphics(filenames[[7]])
```

```{r fig.width=6, fig.height=6}
filenames[[8]] <- here("01_figures", "08_pcaplot.png")

p <- PCAPlot(object = obj, dim.1 = 1, dim.2 = 2, do.return = TRUE)

ggsave(filename = filenames[[8]], width = 6, height = 6, dpi = dpi, units = "in")

knitr::include_graphics(filenames[[8]])
```

```{r fig.width=18, fig.height=24}
filenames[[9]] <- here("01_figures", "09_pcaheatmap.png")

p <- PCHeatmap(
    object = obj,
    pc.use = 1:12,
    cells.use = 500,
    do.balanced = TRUE,
    label.columns = FALSE,
    do.return = TRUE
)

ggsave(filename = filenames[[9]], width = 6, height = 6, dpi = dpi, units = "in")

knitr::include_graphics(filenames[[9]])
```

```{r fig.height=6, fig.width=12}
filenames[[10]] <- here("01_figures", "10_pcablowplot.png")

png(filename = filenames[[10]], width = 18, height = 6, res = dpi, units = "in")
PCElbowPlot(object = obj, num.pc = 100)
dev.off()

knitr::include_graphics(filenames[[10]])
```

```{r}
n.pcs = 75

perform_cluster <- function(obj, n.pcs = 75, res = 0.6, umap=TRUE) {
    obj <- FindClusters(
        object = obj,
        reduction.type = "pca",
        dims.use = 1:n.pcs,
        resolution = res,
        print.output = 0,
        save.SNN = TRUE
    )
    
    obj <- RunTSNE(
        object = obj,
        reduction.use = "pca",
        dims.use = 1:n.pcs,
        max_iter = 2000,
        check_duplicates = FALSE,
        reduction.name = "tSNE",
        reduction.key = "tSNE_"
    )
    
    obj@meta.data <- cbind(obj@meta.data, obj@dr$tSNE@cell.embeddings[rownames(obj@meta.data),])
    
    if (umap) {
        obj <- RunUMAP(
            object = obj,
            reduction.use = "pca",
            dims.use = 1:n.pcs,
            min_dist = 0.75
        )
    
        obj@meta.data <- cbind(obj@meta.data, obj@dr$umap@cell.embeddings[rownames(obj@meta.data),])
  
    }
  
    return(obj)
}

# saveRDS(obj, here("00_data_ingest", "04_rds_generated", "seurat_obj_pcs75_res0.6.rds"))

```

```{r fig.height=6, fig.width=12, ecoh=TRUE, include=TRUE}

p1 <- DimPlot(
    object = obj,
    reduction.use = "tSNE",
    no.legend = TRUE,
    do.return = TRUE,
    vector.friendly = TRUE,
    pt.size = 0.1,
    do.label = TRUE,
    label.size = 10
) +
    ggtitle("tSNE") +
    theme(plot.title = element_text(hjust = 0.5))

p2 <- DimPlot(
    object = obj,
    reduction.use = "umap",
    no.legend = TRUE,
    do.return = TRUE,
    vector.friendly = TRUE,
    pt.size = 0.1,
    do.label = TRUE,
    label.size = 10
    ) +
    ggtitle("UMAP") +
    theme(plot.title = element_text(hjust = 0.5))


filenames[[11]] = here("01_figures", "11_tSNE_UMAP.png")
png(filenames[[11]], height = 20, width = 10, res = dpi, units = "in")
plot_grid(p1, p2, nrow = 2, ncol = 1)
dev.off()

knitr::include_graphics(filenames[[11]])
```


---

### tSNE by Batch
主要通过该图看一下批次效应的矫正情况

```{r include=TRUE}
filenames[["batch"]] = here("01_figures", "tSNE_by_batch.png")

p_list = list()
p_list[["total"]] <- DimPlot(
    object = obj, 
    reduction.use = "tSNE", 
    group.by = "Batch", 
    pt.size = 0.01, 
    do.return = TRUE
)

for (i in unique(obj@meta.data$Batch)) {
    p_list[[as.character(i)]] = DimPlot(
        object = obj,
        reduction.use = "tSNE",
        pt.size = 0.01,
        cells.use = rownames(obj@meta.data[obj@meta.data$Batch == i,]),
        do.return = TRUE,
        do.label = TRUE,
        no.legend = TRUE,
        plot.title = as.character(i)
    )
}

png(filename = filenames[["batch"]], res = dpi, width = 12, height = 12, units = "in")
grid.arrange(grobs = p_list, nrow = 2, ncol = 2)
dev.off()

knitr::include_graphics(filenames[["batch"]])
```

### tSNE by Patient
```{r include=TRUE}
filenames[["patient"]] = here("01_figures", "tSNE_by_patient.png")

p_list = list()
p_list[["total"]] <- DimPlot(
    object = obj, 
    reduction.use = "tSNE", 
    group.by = "PatientID", 
    pt.size = 0.01, 
    do.return = TRUE
)

# for (i in unique(obj@meta.data$PatientID)) {
#     p_list[[as.character(i)]] = DimPlot(
#         object = obj,
#         reduction.use = "tSNE",
#         pt.size = 0.01,
#         cells.use = rownames(obj@meta.data[obj@meta.data$PatientID == i,]),
#         do.return = TRUE,
#         do.label = TRUE,
#         no.legend = TRUE,
#         plot.title = as.character(i)
#     )
# }

png(filename = filenames[["patient"]], width = 10, height = 6, units = "in", res = dpi)
grid.arrange(grobs = p_list)
dev.off()

knitr::include_graphics(filenames[["patient"]])
```


---


```{r}
obj@meta.data$cluster_id <- obj@meta.data$res.0.6

# set the cell_names of obj@meta.data
# :param obj: seurat object
# :param cell_name: string, cell name, if cell_name == "" then use cluser_id as new_cluster_id
# :param cluster_id: vector, contains the target cluster id
# :return : seurat obj
set_cell_names <- function(obj, cell_name, cluster_id) {
    
    obj@meta.data$cell_name <- as.character(obj@meta.data$cluster_id)
    
    if (cell_name != "") {
        new.cluster.ids <- rep(NA, length(unique(obj@meta.data$cluster_id)))
        for (i in cluster_id) {
            new.cluster.ids[i + 1] = cell_name
        }
        
        if (!"cluster_id" %in% colnames(obj@meta.data)) {
            obj@meta.data <- cbind(obj@meta.data, as.data.frame(obj@ident))
            colnames(obj@meta.data)[ncol(obj@meta.data)] = "cluster_id"
        }
    } else {
        new.cluster.ids = cluster_id
    }
    
    if (length(cluster_id) != 0 && length(new.cluster.ids) != length(unique(obj@meta.data$cluster_id))) {
        stop("cluster_id needs equal with obj@meta.data")
    }
    
    
    if (length(cluster_id) == 0) {
        obj@meta.data$cell_name = NA
    } else {
            
        for (i in 1:length(new.cluster.ids)) {
            obj@meta.data$cell_name[obj@meta.data$cell_name == as.character(i - 1)] <- new.cluster.ids[i]
        }
    }

    return(obj)
}


# this function is used to make dimplot without NA, and labeled by cell names
make_dimplot_without_na <- function(obj, no.legend=FALSE, limit.cells = TRUE, umap=FALSE, xlimit=40, ylimit=40) {
    
    used.cells = rownames(obj@meta.data)
    if (limit.cells) {
        used.cells = rownames(obj@meta.data[!is.na(obj@meta.data$cell_name),])
    }
    
    reduction = "tSNE"
    
    if (umap) {
        reduction = "umap"
    }
    
    p <- DimPlot(
      object = obj,
      reduction.use = reduction,
      pt.size = 0.01,
      do.return = TRUE,
      group.by = "cell_name",
      do.label = TRUE,
      no.legend = no.legend,
      cells.use = used.cells
    )
    
    p <- p + coord_cartesian(xlim = c(-1 * xlimit, xlimit), ylim = c(-1 * ylimit, ylimit)) 
    
    return(p)
}


# This function is used to make combination plots, including feature plots, two tSNE plots and dot plot
# :param obj: seurat obj
# :param markers: a vector of string, the marker genes
# :param cell_name: string
# :param cluster_id: a vector of int
make_plots <- function(obj, markers, cell_name, cluster_id, filename, width, height) {
    dim_plot <- FeaturePlot(
        object = obj,
        features.plot = markers,
        cols.use = c(
            "grey",
            "blue"
        ),
        reduction.use = "tSNE",
        do.return = TRUE
    )
    
    
    obj <- set_cell_names(obj, cell_name = cell_name, cluster_id = cluster_id)
    
    if (sum(!is.na(obj@meta.data$cell_name)) > 0) {
        dim_plot[["tSNE"]] = make_dimplot_without_na(obj, limit.cells = FALSE, no.legend = TRUE)
        dim_plot[["tSNE_null"]] = make_dimplot_without_na(obj, limit.cells = TRUE, no.legend = TRUE)
    }
    
    dim_plot[["dotplot"]] = DotPlot(obj, genes.plot = rev(markers), x.lab.rot = T, plot.legend = T, dot.scale = 8, do.return = TRUE)
    
    png(filename, width = width, height = height, res = dpi, units = "in")
    grid.arrange(grobs=dim_plot, nrow = ceiling(length(dim_plot) / 3), ncol = 3)
    dev.off()
}



# function to calculate percentage by two select elements
# :param object: seurat obj
# :param first_ele: str, first select element, main elements
# :param second_ele: str, second select element
# :return dataframe of three columns, var1 -> first element, var2 -> second element, value -> percentage
make_perc_plot <- function(object, first_ele, second_ele) {
    data <- object@meta.data[, c("cluster_id", first_ele, second_ele)] 
    
    res = data.frame("var1"=NA, "var2"=NA, "value"=NA)
    for (i in unique(data[, first_ele])) {
        temp <- data[data[, first_ele] == i, ]
        
        uniq_second_ele = unique(as.vector(data[, second_ele]))
        temp_res = as.data.frame(matrix(NA, nrow = length(uniq_second_ele), ncol=3))

        for (j in 1:length(uniq_second_ele)) {
            temp_res[j, ] <- c(i, uniq_second_ele[j], sum(data[, second_ele] == uniq_second_ele[j] & data[, first_ele] == i) / nrow(temp)) 
        }
        
        colnames(temp_res) <- colnames(res)
        
        res <- rbind(temp_res, res)
    }
    
    res <- na.omit(res)
    res <- res[res[,3] > 0,]
    res[, 3] <- as.numeric(res[, 3])
    return(res)
}
```


## marker genes

以下为点图，用来表示不同细胞亚群中marker基因的表达和分布情况

用以辅助分辨细胞种类

```{r fig.height=8, fig.width=20, include=FALSE}
all_markers <- read.xlsx(here("00_data_ingest", "03_annotation_csv", "gene_list.xlsx"), sheet = 3)

# Airway goblet cells -> 51
# Alveolar Macrophages -> 6*, 35, 49
# Basal -> 5, 34, 45
# B cells -> 13, 23, 46
# CD4+ T cells -> 0, 14, 16, 17, 33, 42
# Clara -> 18, 24
# Club -> 9
# Dentritic -> 30
# Endothelial -> 31
# Erythroid-like and erythroid precursor cells -> 43, 44
# Fibroblasts -> 26, 47
# Mast -> 32
# Monocytes -> 1, 8, 10, 11, 21, 22, 27, 37, 50, 54
# NK -> 3, 25, 53
# NKT -> 2, 7, 12, 19
# Platelets -> 40, 48
# Pulmonary alveolar type I cells -> 38, 39, 41, 55
# Pulmonary alveolar type II cells -> 15, 20, 28, 29*， 36* 
# Treg -> 4

new.cluster.ids <- c(
    "CD4+ T cells", "Monocytes", "NKT", "NK", "Treg",   # 0-4
    "Basal", "Alveolar Macrophages*", "NKT", "Monocytes", "Club cells",              # 4-9
    "Monocytes", "Monocytes", "NKT", "B cells", "CD4+ T cells",        # 10-14
    "Pulmonary alveolar type II cells", "CD4+ T cells", "CD4+ T cells", "Clara", "NKT",      # 15-19
    "Pulmonary alveolar type II cells", "Monocytes", "Monocytes", "B cells", "Clara",      # 20-24
    "NK", "Fibroblasts", "Monocytes", "Pulmonary alveolar type II cells", "Pulmonary alveolar type II cells**",                       # 24-29
    "Dentritic", "Endothelial", "Mast", "CD4+ T cells", "Basal",    # 30-34
    "Alveolar Macrophages", "Pulmonary alveolar type II cells**", "Monocytes", "Pulmonary alveolar type I cells", "Pulmonary alveolar type I cells",    # 35-39
    "Platelets**", "Pulmonary alveolar type I cells", "CD4+ T cells", "Erythroid-like and erythroid precursor cells", "Erythroid-like and erythroid precursor cells",      # 40-44
    "Basal", "B cells", "Fibroblasts", "Platelets**", "Alveolar Macrophages",       # 45-49
    "Monocytes", "Airway goblet cells", "Basal", "NK", "Monocytes",      # 50-54
    "Pulmonary alveolar type I cells"   # 55
)

# obj@meta.data$cell_name <- plyr::mapvalues(x = obj@meta.data$cell_name, from = current.cluster.ids, to = new.cluster.ids)
obj <- set_cell_names(obj, cell_name = "", cluster_id = new.cluster.ids)
# obj@meta.data$cell_name = obj@meta.data$cluster_id

unique_id <- unique(unique(new.cluster.ids))
selected_cols <- rep(NA, length(new.cluster.ids))
for (i in 1:length(unique_id)) {
    selected_cols[which(new.cluster.ids == unique_id[i])] <- custom_colors[i]
}


markers <- unique(all_markers$Markers)

p <- SplitDotPlotGG(
    obj,
    genes.plot = rev(markers),
    x.lab.rot = T,
    plot.legend = F,
    dot.scale = 8,
    do.return = T,
    grouping.var = "cluster_id",
    cols.use = selected_cols,
    group.by = "cell_name"
)

filenames[["splitDot"]] <- here("01_figures", "splitDotPlot.png")
ggsave(filenames[["splitDot"]], plot = p, width = 25, height = 15, dpi = dpi, units = "in")
```
```{r}
knitr::include_graphics(filenames[["splitDot"]])
```

- 点的大小代表该细胞亚群的细胞中有该基因表达的比例；
- x轴为测试所用marker基因
- y轴为Suerat划分的亚群以及目前该亚群可能所属的细胞种类
- 其中带有`*`后缀的为通过后续的cluster tree识别出的细胞

```{r include=FALSE}
p <- DotPlot(
    obj, 
    genes.plot = rev(markers), 
    dot.scale = 8, 
    x.lab.rot = T, 
    group.by = "cell_name", 
    plot.legend = T,
    do.return = TRUE
)

filenames[["dotplot"]] <- here("01_figures", "Dotplot.png")
ggsave(filenames[["dotplot"]], plot = p, width = 25, height = 20, dpi = dpi, units = "in")
```
```{r}
knitr::include_graphics(filenames[["dotplot"]])
```

- 点的大小代表该细胞亚群的细胞中有该基因表达的比例；
- 点的颜色深浅代表该基因在该细胞亚群所有细胞中的平均表达水平；
- x轴为测试所用marker基因
- y轴为该亚群可能所属的细胞种类
---
## Make percentage plots

以下为每个识别的细胞亚群中来自不同患者或者不同批次1（华西自测的样本）、2（上海公司所测）以及3（Nature Medicine文章中的数据）

1. make percentage plot by cluster id
```{r}

## make plot of cluster - patient
perc_by_patient <- make_perc_plot(obj, "cluster_id", "PatientID")
p1 <- ggplot(perc_by_patient, aes(x=var1, y=value, fill=var2)) + 
    geom_bar(stat = "identity") +
    labs(x="cluster", y = "percentage of cells") +
    guides(fill=guide_legend(title="Patient", ncol = 20)) +
    theme(
        axis.title.x = element_blank(), 
        axis.text.x = element_blank(),
        legend.position = "top"
    )

## make plot of cluster - batch
perc_by_batch <- make_perc_plot(obj, "cluster_id", "Batch")
p2 <- ggplot(perc_by_batch, aes(x=var1, y=value, fill=var2)) + 
    geom_bar(stat = "identity") +
    labs(x="cluster", y = "percentage of cells") +
    guides(fill=guide_legend(title="Batch")) +
     theme(
        legend.position = "bottom"
    )

filenames[["percent"]] = here("01_figures", "percentage_by_meta.png")

png(filename = filenames[["percent"]], width = 20, height = 12, units = "in", res = dpi)
grid.newpage()
grid.draw(rbind(ggplotGrob(p1), ggplotGrob(p2), size = "last"))
dev.off()

knitr::include_graphics(filenames[["percent"]])
```

2. make percentaget plot by cell_name
```{r}

## make plot of cluster - patient
perc_by_patient <- make_perc_plot(obj, "cell_name", "PatientID")
p1 <- ggplot(perc_by_patient, aes(x=var1, y=value, fill=var2)) + 
    geom_bar(stat = "identity") +
    labs(x="cluster", y = "percentage of cells") +
    guides(fill=guide_legend(title="Patient", ncol = 20)) +
    theme(
        axis.title.x = element_blank(), 
        axis.text.x = element_blank(),
        legend.position = "top"
    )

## make plot of cluster - batch
perc_by_batch <- make_perc_plot(obj, "cell_name", "Batch")
p2 <- ggplot(perc_by_batch, aes(x=var1, y=value, fill=var2)) + 
    geom_bar(stat = "identity") +
    labs(x="cluster", y = "percentage of cells") +
    guides(fill=guide_legend(title="Batch")) +
     theme(
        legend.position = "bottom",
        axis.text.x = element_text(angle = 45, hjust = 1)
    )

filenames[["percent_cell"]] = here("01_figures", "percentage_by_cell_name.png")

png(filename = filenames[["percent_cell"]], width = 20, height = 12, units = "in", res = dpi)
grid.newpage()
grid.draw(rbind(ggplotGrob(p1), ggplotGrob(p2), size = "last"))
dev.off()

knitr::include_graphics(filenames[["percent_cell"]])
```

```{r}

obj@meta.data$cluster_name <- paste(obj@meta.data$cell_name, obj@meta.data$cluster_id, sep = "_")

## make plot of cluster - patient
perc_by_patient <- make_perc_plot(obj, "cluster_name", "PatientID")
p1 <- ggplot(perc_by_patient, aes(x=var1, y=value, fill=var2)) + 
    geom_bar(stat = "identity") +
    labs(x="cluster", y = "percentage of cells") +
    guides(fill=guide_legend(title="Patient", ncol = 20)) +
    theme(
        axis.title.x = element_blank(), 
        axis.text.x = element_blank(),
        legend.position = "top"
    )

## make plot of cluster - batch
perc_by_batch <- make_perc_plot(obj, "cluster_name", "Batch")
p2 <- ggplot(perc_by_batch, aes(x=var1, y=value, fill=var2)) + 
    geom_bar(stat = "identity") +
    labs(x="cluster", y = "percentage of cells") +
    guides(fill=guide_legend(title="Batch")) +
     theme(
        legend.position = "bottom",
        axis.text.x = element_text(angle = 45, hjust = 1)
    )

filenames[["percent_xxx"]] = here("01_figures", "percentage_by_cluster_name.png")

png(filename = filenames[["percent_xxx"]], width = 20, height = 12, units = "in", res = dpi)
grid.newpage()
grid.draw(rbind(ggplotGrob(p1), ggplotGrob(p2), size = "last"))
dev.off()

knitr::include_graphics(filenames[["percent_xxx"]])
```

```{r fig.height=8, fig.width=12}
make_dimplot_without_na(obj, limit.cells = FALSE)
```

### Suerat系统进化树

借助该树辅助判断分辨的细胞亚群是否可靠

```{r}
# # 经过拆解代码发现，该functions底层调用ape::as.phylo来计算进化树 （代码/R/cluster_determination.R）
# # 
# obj <- BuildClusterTree(
#     object=obj, 
#     genes.use = all_markers, 
#     pcs.use = 1:n.pcs, 
#     SNN.use = NULL,
#     do.plot = TRUE, 
#     do.reorder = FALSE, 
#     reorder.numeric = FALSE,
#     show.progress = TRUE,
#     do.plot = FALSE
# )


filenames[["clusterTree"]] <- here("01_figures", "cluster_tree.png")

png(filename = filenames[["clusterTree"]], width = 20, height = 15, res = dpi, units = "in")
PlotClusterTree(object = obj)
dev.off()

knitr::include_graphics(filenames[["clusterTree"]])
```

> Seurat默认效果图，下图的效果会更好些


```{r}
# PlotClusterTree(obj)

tree <- obj@cluster.tree[[1]]

for(i in 1:length(new.cluster.ids)) {
    cluster_id <- paste("cluster_id", i - 1, sep = ":")
    cell_num <- paste("cell_num", sum(obj@meta.data$cluster_id == as.character(i - 1)), sep = ":")
    
    final <- paste(new.cluster.ids[i], cluster_id, sep = "; ")
    final <- paste(final, cell_num, sep = "; ")
    
    tree$tip.label[tree$tip.label == as.character(i - 1)] <- final
}


filenames[["new_cluster_tree"]] = here("01_figures", "cluster_tree_by_cell.png")

png(filename = filenames[["new_cluster_tree"]], width = 20, height = 12, units = "in", res = dpi)
plot.phylo(x = tree, direction = "downwards")
nodelabels()
dev.off()
knitr::include_graphics(filenames[["new_cluster_tree"]])

```
> 综合多种信息，包括目前识别的细胞种类，原cluster_id以及每个cluster中的细胞数量。细胞数量过少可能会影响到该树的构建

---
## feature plots
使用


---
###  Airway goblet cells
```{r include=FALSE}
# https://panglaodb.se/markers.html?cell_type=%27Platelets%27

filename <- here("02_feature_plots", "Airway_goblet_cells.png")

markers <- c(
    "SCGB3A1",
    "TFF2",
    "PIGR",
    "CEACAM1"
)

make_plots(
    obj = obj,
    markers = markers,
    cell_name = "Airway goblet cells", 
    cluster_id = c(51),
    filename = filename,
    width = cols * size, 
    height = ceiling((length(markers) + 3) / cols) * size
)
```
```{r}
knitr::include_graphics(filename)
```


---
### Alveolar Macrophages
```{r include=FALSE}
filename <- here("02_feature_plots", "Alveolar_Macrophagess.png")

markers <- c(
    "ALDH2",
    "APOBEC3A",
    "C15orf48",
    "C1QC",
    "CCL8",
    "EEF1A1",
    "F13A1",
    "FOLR2",
    "IFIT1",
    "CD14",    # Monocytes, Macrophages,   used
    "CD68",    # dendritic, Interstitial macrophages, monocytes
    "CSF1R",   # dendritic, Interstitial macrophages, monocytes
    "ITGAX",  # Dendritic, alveolar macrophages
    "MARCO",  # Dendritic, alveolar macrophages
    "MRC1",   # alveolar macrophages, Monocytes
    "S100A4",
    "TNFAIP2",
    "CLEC7A",
    "CCL3"
)

make_plots(
    obj = obj,
    markers = markers,
    cell_name = "Macrophages", 
    cluster_id = c(6, 35, 49), 
    filename = filename,
    width = cols * size, 
    height = ceiling((length(markers) + 3) / cols) * size
)
```
```{r}
knitr::include_graphics(filename)
```


---
### Basal
```{r include=FALSE}

filename <- here("02_feature_plots", "Basal.png")

markers <- c(
    "KRT5",
    # "P63",
    # "IGTA6", 
    "NGFR", 
    "KRT14"
)

make_plots(
    obj = obj,
    markers = markers,
    cell_name = "Basal cells", 
    cluster_id = c(5, 34, 45),
    filename = filename,
    width = cols * size, 
    height = ceiling((length(markers) + 3) / cols) * size
)
```
```{r}
knitr::include_graphics(filename)
```

---
### B cells
```{r include=FALSE}
filename <- here("02_feature_plots", "B_cell.png")

markers <- c(
    "MS4A1",    # B
    "CD79A",
    "CD19"
)

make_plots(
    obj = obj,
    markers = markers,
    cell_name = "B cells", 
    cluster_id = c(13, 23, 46),
    filename = filename,
    width = cols * size, 
    height = ceiling((length(markers) + 3) / cols) * size
)
```
```{r}
knitr::include_graphics(filename)
```

---
4. CD4+ T cells
```{r include=FALSE}
filename <- here("02_feature_plots", "CD4_T_cell.png")

markers <- c(
    "CD2",
    "CD3D",
    "CD3E",
    "CD3G",
    "TRAC",
    "CD4",
    "IL7R",
    "CD44"
)

make_plots(
    obj = obj,
    markers = markers,
    cell_name = "T cells", 
    cluster_id = c(0, 14, 16, 17, 33, 42),
    filename = filename,
    width = cols * size, 
    height = ceiling((length(markers) + 3) / cols) * size
)
```

```{r}
knitr::include_graphics(filename)
```

---
### CD8+ T cells
```{r include=FALSE}
filename <- here("02_feature_plots", "CD8_T_cell.png")

markers <- c(
    "CD8A",
    "CD8B",
    "CCL4L2",
    "GZMK",
    "LEF1",
    "LAYN"
)

make_plots(
    obj = obj,
    markers = markers,
    cell_name = "CD8+ T cells", 
    cluster_id = c(),
    filename = filename,
    width = cols * size, 
    height = ceiling((length(markers) + 3) / cols) * size
)
```
```{r}
knitr::include_graphics(filename)
```


---
### Clara cells
```{r include=FALSE}
# https://panglaodb.se/markers.html?cell_type=%27Platelets%27

filename <- here("02_feature_plots", "Clara_cells.png")

markers <- c(
    # "CBR2",
    "SCGB1A1",
    "SFTPB",
    "BPIFA1",
    "SCGB3A2",
    "MUC5B"
)

make_plots(
    obj = obj,
    markers = markers,
    cell_name = "Clara cells", 
    cluster_id = c(18, 24),
    filename = filename,
    width = cols * size, 
    height = ceiling((length(markers) + 3) / cols) * size
)
```
```{r}
knitr::include_graphics(filename)
```

---
### Club
```{r include=FALSE}
filename <- here("02_feature_plots", "Club.png")

markers <- c(
    "SCGB1A1",
    "BPIFB1",
    "CXCL1",
    "CP",
    "PIGR",
    "MUC5B", 
    "SAA2",
    "TSPAN8",
    "FOXJ1",   # ciliated
    "DNAI1",    # ciliated
    "SEC14L3",
    "TMEM212"
)

make_plots(
    obj = obj,
    markers = markers,
    cell_name = "Club cells", 
    cluster_id = c(9),
    filename = filename,
    width = cols * size, 
    height = ceiling((length(markers) + 3) / cols) * size
)
```
```{r}
knitr::include_graphics(filename)
```

---
### Dendritic Cells
```{r include=FALSE}
filename <- here("02_feature_plots", "Dendritic.png")

markers <- c(
    "FCER1A",
    "C1orf54",
    "CCR7",
    "LGALS2",
    "LSP1",
    "NAAA",
    "CCL17",
    "CD68",    # dendritic, Interstitial macrophages, monocytes
    "COL1A1",  # Stromal cells
    "CSF1R",   # dendritic, Interstitial macrophages, monocytes
    "ITGAX",   # Dendritic, alveolar macrophages
    "MARCO"    # Dendritic, alveolar macrophages
)

make_plots(
    obj = obj,
    markers = markers,
    cell_name = "Dendritic", 
    cluster_id = c(30),
    filename = filename,
    width = cols * size, 
    height = ceiling((length(markers) + 3) / cols) * size
)
```
```{r}
knitr::include_graphics(filename)
```


---
### Endothelial cells
```{r include=FALSE}
filename <- here("02_feature_plots", "Endothelial.png")

markers <- c(
    "ADM",     # normal
    "AKAP12",
    "BIRC3",
    "CLEC2B",
    "IL1RL1",
    "SLC2A3",  # lymphatic
    "MMRN1",
    "IGFBP3",   # tumor
    "SPRY1",
    "ENPP2"
)


make_plots(
    obj = obj,
    markers = markers,
    cell_name = "Endothelial cells", 
    cluster_id = c(31, 47),
    filename = filename,
    width = cols * size, 
    height = ceiling((length(markers) + 3) / cols) * size
)
```
```{r}
knitr::include_graphics(filename)
```


---
### Erythroblasts
```{r include=FALSE}
# https://panglaodb.se/markers.html?cell_type=%27Platelets%27

filename <- here("02_feature_plots", "Erythroblasts.png")

markers <- c(
    "BCL11A",
    "CDH1",
    "TFRC",
    # "KLF1",
    "WRN",
    "CGA"
)

make_plots(
    obj = obj,
    markers = markers,
    cell_name = "Erythroblasts", 
    cluster_id = c(31),
    filename = filename,
    width = cols * size, 
    height = ceiling((length(markers) + 3) / cols) * size
)
```
```{r}
knitr::include_graphics(filename)
```


---
### Erythroid-like and erythroid precursor cells
```{r include=FALSE}
filename <- here("02_feature_plots", "Erythrocyte_like_and_erythroid_precursor_cells.png")

markers <- c(
    # "SLC4A1",    # CD233
    # "GYPA",     # CD235A
    "HBB",       # Erythroid-like and erythroid precursor cells
    # "ALAS2"
    "HBA1",
    # "HBD",
    "SLC25A37",
    "SLC25A39",
    "SNCA",
    "DCAF12"
)

make_plots(
    obj = obj,
    markers = markers,
    cell_name = "Erythroid-like and erythroid precursor cellse", 
    cluster_id = c(43, 44),
    filename = filename,
    width = cols * size, 
    height = ceiling((length(markers) + 3) / cols) * size
)
```
```{r}
knitr::include_graphics(filename)
```


---
### Fibroblasts
```{r include=FALSE}
filename <- here("02_feature_plots", "Fibroblasts.png")

markers <- c(
    "ACTG2",
    "ALDH1A3",
    "ASPN",
    "C3",
    "COL10A1",
    "COL13A1",   # 在 《Single-Cell Deconvolution of FibroblastHeterogeneity in Mouse Pulmonary Fibrosis》文章中小鼠分别以COL13A1和COL14A1为标记，分到了两个fibroblasts群
    "COL14A1",
    "COL18A1",
    "DKK1",
    # "HIGD1B"
    # "LHFP"
    "HTRA3",
    "NFKBIA",
    "PTN",
    "PTX3",
    "TNFRSF12A",
    "COL1A1"  # Stromal cells
)

make_plots(
    obj = obj,
    markers = markers,
    cell_name = "Fibroblasts", 
    cluster_id = c(26, 47),
    filename = filename,
    width = cols * size, 
    height = ceiling((length(markers) + 3) / cols) * size
)
```
```{r}
knitr::include_graphics(filename)
```


---
### Granulocyte
```{r include=FALSE}
filename <- here("02_feature_plots", "Granulocyte.png")

markers <- c(
    "ANPEP",    # CD13
    "FUT4",     # CD15
    "CD33",
    "CD300E",
    "CXCL8",
    "EREG"
)

make_plots(
    obj = obj,
    markers = markers,
    cell_name = "Granulocytes", 
    cluster_id = c(),
    filename = filename,
    width = cols * size, 
    height = ceiling((length(markers) + 3) / cols) * size
)
```
```{r}
knitr::include_graphics(filename)
```


---
### Mast
```{r include=FALSE}
filename <- here("02_feature_plots", "Mast.png")

markers <- c(
    "AREG",    # mast cells one of Granulocyte
    "CLU",     # mast cells
    "MS4A2",   # mast cells
    "PTGS1",
    "RGS13"
)

make_plots(
    obj = obj,
    markers = markers,
    cell_name = "Mast", 
    cluster_id = c(32),
    filename = filename,
    width = cols * size, 
    height = ceiling((length(markers) + 3) / cols) * size
)
```
```{r}
knitr::include_graphics(filename)
```


---
### Monocytes
```{r include=FALSE}
filename <- here("02_feature_plots", "Monocytes.png")

markers <- c(
    "LYZ",
    "CD33", 
    "FCGR1A", 
    "CD14",    # Monocytes, Macrophages,   used
    "CD68",    # dendritic, Interstitial macrophages, monocytes
    "CSF1R",   # dendritic, Interstitial macrophages, monocytes
    "ITGAX",  # Dendritic, alveolar macrophages
    "MARCO",  # Dendritic, alveolar macrophages
    "MRC1",   # alveolar macrophages, Monocytes
    "FCGR3A"
)

make_plots(
    obj = obj,
    markers = markers,
    cell_name = "Monocytes", 
    cluster_id = c(1, 8, 10, 11, 21, 22, 27, 37, 50, 54),
    filename = filename,
    width = cols * size, 
    height = ceiling((length(markers) + 3) / cols) * size
)
```
```{r}
knitr::include_graphics(filename)
```


---
### NK
```{r include=FALSE}
filename <- here("02_feature_plots", "NK_cell.png")

markers <- c(
    "GNLY",
    "NKG7"
)

make_plots(
    obj = obj,
    markers = markers,
    cell_name = "NKT cells", 
    cluster_id = c(3, 25, 53, 2, 7, 12, 19),
    filename = filename,
    width = cols * size, 
    height = ceiling((length(markers) + 3) / cols) * size
)
```
```{r}
knitr::include_graphics(filename)
```


---
###. Platelets
```{r include=FALSE}
# https://panglaodb.se/markers.html?cell_type=%27Platelets%27

filename <- here("02_feature_plots", "Platelets.png")

markers <- c(
    "DAB2",
    # "PFN1P1",
    "PECAM1",
    "BIN2",
    "PLEK",
    "CXCR4",
    "FERMT3",
    "MMD",
    "PRKAR2B"
    # "P2RY12",
    # "PF4"
)

make_plots(
    obj = obj,
    markers = markers,
    cell_name = "Platelets", 
    cluster_id = c(40, 48),
    filename = filename,
    width = cols * size, 
    height = ceiling((length(markers) + 3) / cols) * size
)
```
```{r}
knitr::include_graphics(filename)
```



---
### Pulmonary alveolar type I cells
```{r include=FALSE}
# https://panglaodb.se/markers.html?cell_type=%27Platelets%27

filename <- here("02_feature_plots", "Pulmonary_alveolar_type_I_cells.png")

markers <- c(
    "HOPX",
    "AQP5",
    "CLDN18",
    "AGER",
    "BTG3",
    "RNF5",
    "ACOT9",
    "CTGF",
    "LMO7",
    "LAMB2",
    "SPHK1",
    "BMP4"
)

make_plots(
    obj = obj,
    markers = markers,
    cell_name = "Clara cells", 
    cluster_id = c(38, 39, 41, 55),
    filename = filename,
    width = cols * size, 
    height = ceiling((length(markers) + 3) / cols) * size
)
```
```{r}
knitr::include_graphics(filename)
```


---
### Pulmonary alveolar type II cells
```{r include=FALSE}
# https://panglaodb.se/markers.html?cell_type=%27Platelets%27

filename <- here("02_feature_plots", "Pulmonary_alveolar_type_II_cells.png")

markers <- c(
    "NAPSA",
    "LPCAT1",
    # "CBR2",
    "ETV5",
    "SFTPC",
    "SFTPB",
    "SFTPA1",
    "AGER",
    "CLDN18",
    # "CXCL15",
    "SFTPD",
    "LAMP3",
    "SLC34A2",
    "ABCA3"
)

make_plots(
    obj = obj,
    markers = markers,
    cell_name = "Clara cells", 
    cluster_id = c(15, 20, 28, 29, 36),
    filename = filename,
    width = cols * size, 
    height = ceiling((length(markers) + 3) / cols) * size
)
```
```{r}
knitr::include_graphics(filename)
```



---
### Treg
```{r include=FALSE}
filename <- here("02_feature_plots", "Treg.png")

markers <- c(
    "TNFRSF4",    # Treg
    "FOXP3",    # Treg
    "IL2RA",   # Treg
    "CTLA4",    # Treg
    "LTB",
    "AC133644.2"
)

make_plots(
    obj = obj,
    markers = markers,
    cell_name = "Treg", 
    cluster_id = c(4),
    filename = filename,
    width = cols * size, 
    height = ceiling((length(markers) + 3) / cols) * size
)
```
```{r}
knitr::include_graphics(filename)
```




<!-- 11. Megakaryocytes -->
<!-- ```{r fig.height=32, fig.width=24, include=FALSE} -->
<!-- filename <- here("02_feature_plots", "12_Megakaryocytes.png") -->

<!-- markers <- c( -->
<!--     "PPBP", -->
<!--     # "ITGA2B", # CD41 -->
<!--     # "GP1BA",  # CD42A -->
<!--     # "GP1BB",  # CD42B -->
<!--     "ITGAV"     # CD51 -->
<!--     # "ITGB3",  # CD61 -->
<!--     # "PF4", -->
<!--     # "GP9" -->
<!-- ) -->

<!-- make_plots( -->
<!--     obj = obj, -->
<!--     markers = markers, -->
<!--     cell_name = "Megakaryocytes",  -->
<!--     cluster_id = c(1), -->
<!--     filename = filename, -->
<!--     width = cols * size,  -->
<!--     height = ceiling((length(markers) + 3) / cols) * size -->
<!-- ) -->
<!-- ``` -->
<!-- ```{r} -->
<!-- knitr::include_graphics(filename) -->
<!-- ``` -->













<!-- 22. Ionocytes -->
<!-- ```{r include=FALSE} -->
<!-- # https://panglaodb.se/markers.html?cell_type=%27Platelets%27 -->

<!-- filename <- here("02_feature_plots", "23_Ionocytes.png") -->

<!-- markers <- c( -->
<!--     "CFTR", -->
<!--     "FOXI1", -->
<!--     # "ASCL3", -->
<!--     "SCGB1A1" -->
<!-- ) -->

<!-- make_plots( -->
<!--     obj = obj, -->
<!--     markers = markers, -->
<!--     cell_name = "Clara cells",  -->
<!--     cluster_id = c(1), -->
<!--     filename = filename, -->
<!--     width = cols * size,  -->
<!--     height = ceiling((length(markers) + 3) / cols) * size -->
<!-- ) -->
<!-- ``` -->
<!-- ```{r} -->
<!-- knitr::include_graphics(filename) -->
<!-- ``` -->


---
### Save raw data by different cells

```{r}

obj@meta.data$cell_name <- sapply(obj@meta.data$cell_name, function(x){gsub("\\*", "", x)})

for(i in unique(obj@meta.data$cell_name)) {
    meta <- obj@meta.data[obj@meta.data$cell_name == i, ]
    data <- obj@raw.data[, rownames(meta)]
    
    obj_tiss <- CreateSeuratObject(
        data,
        project = i,
        meta.data = meta
    )
    
    saveRDS(obj_tiss, here("00_data_ingest", "04_rds_generated", "seurat_obj", paste(gsub("\\s", "_", i), ".rds", sep = "")))
}

```


Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
