---
title: "APII"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

root.dir = "LungCancer10x/"

full.path <- function(...) { return(paste(root.dir, ..., sep = "/")) }
```


```{r eval=FALSE, include=FALSE}
library(Seurat)
library(here)
library(scatterpie)
library(dplyr)
library(ggthemes)
library(ggrastr)
library(ComplexHeatmap)
library(circlize)
library(wesanderson)
library(openxlsx)
library(stringr)
# library(Mfuzz)
library(reticulate)
library(ggradar)
library(reshape2)
library(ggrepel)
library(ggpubr)
library(ggthemes)
library(destiny)
library(harmony)
library(UBL)

extrafont::loadfonts()
```


```{R}
# obj <- readRDS(full.path("03_each_cells/Batch/ATII/LUAD/seurat.rds"))
obj <- readRDS(full.path("11_CNV/each_cells/ATII/LUAD/seurat_obj.rds"))
dm <- readRDS(full.path("11_CNV/each_cells/ATII/LUAD/URD.rds"))
```


```{r}
gg_color_hue <- function(n) {
    hues = seq(15, 375, length = n + 1)
    hcl(h = hues, l = 65, c = 100)[1:n]
}

stage_colors = c("I"="#65A9A3", "II"="#4A933E", "III"="#EC7A21", "IV"="#D73F47", "LUAD_Normal"="#FECC1B", "LUSC_Normal"="#778793")
disease_colors = c("LUAD" = "#0084D1", "LUAD_Normal"="#FECC1B", "Normal"="#73BDFF", "LUSC_Normal"="#778793", "LUSC"="#A0C807", "LELC"="#6A0019", "CPI"="#C5000B")
tissue_colors = c("Bronichi"="#FB290F", "Lymph node"="#488F17", "Tumor"="#FC8210")

cluster_cols = c(
    as.character(wes_palette("Zissou1")),
    as.character(wes_palette("Royal2")),
    as.character(wes_palette("Royal1")),
    as.character(wes_palette("Darjeeling2")),
    as.character(wes_palette("Darjeeling1"))
)

patient_colors = cluster_cols[1:length(unique(obj@meta.data$PatientID))]
names(patient_colors) <- unique(obj@meta.data$PatientID)

colors = c(stage_colors, disease_colors, tissue_colors, patient_colors)

labels_size = 12
title_size = 15
```



```{R}
make_feature_plot <- function(
    obj,
    gene, 
    coord=NULL, 
    pt.size = 0.1, 
    alpha = 0.5, 
    sort = T,
    reduction.use = "umap"
) {
    library(ggplot2)
    library(ggrastr)
    data = as.data.frame(obj@meta.data)
    data$value = obj@data[which(rownames(obj@data) == gene), rownames(data)]
    
    if(is.null(coord)) {
        coord = obj@dr[[reduction.use]]@cell.embeddings
    }
    
    data$x = coord[rownames(data), 1]
    data$y = coord[rownames(data), 2]
    
    if(sort) {
        data = data[order(data$value, decreasing = F), ]
    }
    
    p <- ggplot(data = data, aes(x=x, y=y, color = value)) +
        geom_point_rast(size = pt.size, alpha = alpha) +
        scale_color_gradient(low = "lightgrey", high = "blue") +
        theme_bw(base_family = "Arial Unicode MS") +
        labs(title = gene, x="UMAP1", y="UMAP2") +
        theme(
            plot.title = element_text(hjust = 0.5, face="bold", size = 20), 
            legend.position = "none",
            axis.text = element_text(size = 12),
            axis.title = element_text(size = 15),
            panel.grid = element_blank()
        )
    p
}
```



```{r fig.height=4, fig.width=12}
# Function to make combined bar plots
# 1. percentage plot by Stage
# 2. percentage plot by Disease
# 3. percentage plot by PatientID
# 4. barplot of number of cells
# 5. boxplots of number of transcripts
# Note: custom_colors needs to exists in namespace, cause this function using customed color palette
# :param object: Seurat object 2.3+
# :param group.by: how to make groups
# :param group.type: if int is set, sort the plot order by numeric, else by str
# :param rel_height: this function separate legend and main plot into two grid (up and down), so using this to adjust the relative position of legend and main plot
# :param with_disease: whethter plot the disease percentage
# :param ncol: the legend ncol of patient plot
# :return ggplot2 object
make_combined_barplot <- function(
    object, 
    group.by = "res.0.8", 
    group.type = "int", 
    rel_height = c(0.2, 1),
    with_disease=TRUE,
    ncol=4
) {
    library(scales)
    object@meta.data$PatientID = as.character(object@meta.data$PatientID)
    ## make percentage bar plot with stage
    meta <- object@meta.data
    meta$group <- as.character(meta[, group.by])
    # meta$newTissue <- as.character(meta$Tissue)
    # meta$newTissue[meta$newTissue == "Tumor" & meta$Disease != ""] <- as.character(meta$Disease[meta$newTissue == "Tumor" & meta$Disease != ""])

    plot_order = c(unique(meta$group), "Total")
    
    if(group.type == "int") {
        plot_order = as.character(c(sort(as.numeric(unique(meta$group))), "Total"))
    }
    
    ### make percentage plot of stage
    meta$Stage = as.character(meta$Stage)
    temp <- as.data.frame(
        meta %>% 
            dplyr::select(group, Stage) %>% 
            dplyr::group_by(group, Stage) %>% 
            dplyr::add_tally() %>% 
            unique() %>% 
            dplyr::group_by(group) %>% 
            dplyr::mutate(freq = n / sum(n)) %>% unique()
    )
    
    temp1 <- as.data.frame(table(meta$Stage))
    colnames(temp1) <- c("Stage", "n")
    
    temp1$group = "Total"
    temp1$freq = temp1$n / sum(as.numeric(temp1$n))
    
    temp1 <- temp1[, colnames(temp)]
    
    temp <- rbind(temp, temp1)
    
    temp$group <- factor(temp$group, levels = plot_order)
    
    p1 <- ggplot(data = temp, aes(x=group, y=freq, fill=Stage)) + 
        geom_bar(stat = "identity", width = .75) + 
        coord_flip() + 
        theme_bw() +
        theme(
            axis.text = element_text(size = 15),
            axis.title = element_text(size = 15),
            legend.text = element_text(size = 15),
            legend.title = element_blank(), 
            legend.justification = "center"
        ) +
        labs(y="Fraction of cells", x = "") +
        # scale_x_discrete(breaks=plot_order) + 
        guides(fill = guide_legend(ncol=4)) + 
        scale_fill_manual(values=stage_colors)
    
    p1_legend <- get_legend(p1)
    
    p1 <- p1 + theme(legend.position = "none")
    print("p1")
    
    ### make percentage of tissue
    temp <- as.data.frame(
        meta %>% 
            dplyr::select(group, Disease) %>% 
            dplyr::group_by(group, Disease) %>% 
            dplyr::add_tally() %>% 
            unique() %>% 
            dplyr::group_by(group) %>% 
            dplyr::mutate(freq = n / sum(n)) %>% 
            unique()
    )
    
    temp1 <- as.data.frame(table(meta$Disease))
    
    colnames(temp1) <- c("Disease", "n")
    temp1$group = "Total"
    temp1$freq = temp1$n / sum(as.numeric(temp1$n))
    
    temp1 <- temp1[, colnames(temp)]
    temp <- rbind(temp, temp1)
    
    temp$group <- factor(temp$group, levels = plot_order)
    
    ## combine colors of tissue and disease
    p2 <- ggplot(data = temp, aes(x=group, y=freq, fill=Disease)) + 
        geom_bar(stat = "identity", width = .75) + 
        coord_flip() + 
        theme_bw() +
        theme(
            legend.position = "right", 
            legend.title = element_blank(), 
            legend.justification = "center",
            axis.text = element_text(size = 15),
            axis.title = element_text(size = 15),
            legend.text = element_text(size = 15)
        ) +
        labs(y="Fraction of cells", x = "") +
        # scale_x_discrete(breaks=plot_order) + 
        guides(fill = guide_legend(ncol=1)) + 
        scale_fill_manual(values=colors)
    
    p2_legend <- get_legend(p2)
    
    p2 <- p2 + theme(legend.position = "none")
    print("p2")
    
    ### make percentage barplot of patients
    temp <- as.data.frame(
        meta %>% 
            dplyr::select(group, PatientID) %>% 
            dplyr::group_by(group, PatientID) %>% 
            dplyr::add_tally() %>% 
            unique() %>% 
            dplyr::group_by(group) %>% 
            dplyr::mutate(freq = n / sum(n)) %>% 
            unique()
    )
    
    temp1 <- as.data.frame(table(object@meta.data$PatientID))
    colnames(temp1) <- c("PatientID", "n")
    temp1$group = "Total"
    temp1$freq = temp1$n / sum(temp1$n)
    
    temp <- rbind(temp, temp1)
    temp$group <- factor(temp$group, levels = plot_order)
    
    p3 <- ggplot(data = temp, aes(x=group, y=freq, fill=PatientID)) + 
        geom_bar(stat = "identity", width = .75) + 
        coord_flip() + 
        theme_bw() +
        theme(
            legend.position = "right", 
            legend.title = element_blank(), 
            legend.justification = "center",
            legend.spacing.x = unit(0.1, 'cm'),
            axis.text = element_text(size = 15),
            axis.title = element_text(size = 15),
            legend.text = element_text(size = 10)
        ) +
        labs(y="Fraction of cells", x = "") +
        # scale_x_discrete(breaks=plot_order) + 
        guides(fill = guide_legend(ncol=ncol)) + 
        scale_fill_manual(values=colors)
    p3_legend <- get_legend(p3)
    
    p3 <- p3 + theme(legend.position = "none")
    print("p3")
    
    ### make bar plot of number of cells
    temp <- as.data.frame(
        meta %>% 
            dplyr::select(group) %>% 
            dplyr::group_by(group) %>% 
            dplyr::add_tally() %>% 
            unique()
    )
    
    temp1 <- data.frame(group="Total", n=nrow(meta))
    temp <- rbind(temp, temp1)
    temp$group <- factor(temp$group, levels = plot_order)
    p4 <- ggplot(temp, aes(x=group, y=n, fill=stata_pal("s1color")(1))) + 
        geom_bar(stat = "identity") +
        coord_flip() + 
        theme_bw() +
        theme(
            legend.position = "none",
            axis.text = element_text(size = 15),
            axis.title = element_text(size = 15),
            legend.text = element_text(size = 10)
        ) +
        labs(y="Number of cells", x = "") +
        # scale_x_discrete(breaks=plot_order) +
        scale_y_log10(
            breaks = trans_breaks("log10", function(x) 10^x),
            labels = trans_format("log10", math_format(10^.x))
        )
    print("p4")
    
    ### make boxplot of transcript number
    meta1 = meta
    meta1$group <- "Total"
    meta <- rbind(meta, meta1)
    meta$group <- factor(meta$group, levels = plot_order)

    p5 <- ggplot(meta, aes(x=group, group=group, y=nUMI)) + 
        geom_boxplot(fill = hc_pal()(1)) + 
        coord_flip() + 
        # scale_x_discrete(breaks=plot_order[1:(length(plot_order) - 1)]) +
        labs(y="Number of transcripts", x = "") + 
        scale_y_log10(
            breaks = trans_breaks("log10", function(x) 10^x),
            labels = trans_format("log10", math_format(10^.x))
        ) +
        theme_bw() + 
        theme(
            axis.text = element_text(size = 15),
            axis.title = element_text(size = 15),
            legend.text = element_text(size = 10)
        )
    print("p5")
    
    plots = list(p1_legend, p2_legend, p3_legend, ggplot() + theme_void(), ggplot() + theme_void() , p1, p2, p3, p4, p5)
    
    if (!with_disease) {
        plots = list(p1_legend, p3_legend, ggplot() + theme_void(), ggplot() + theme_void(), p1, p3, p4, p5)
    }
    
    p <- plot_grid(plotlist=plots, nrow = 2, rel_heights = rel_height)
    return(p)
}

```

```{r}
make_module_heatmap <- function(obj, res, mark_genes) {
    
    meta = obj@meta.data[order(obj@meta.data$Stage, obj@meta.data$PatientID), ]
    
    meta$Disease <- as.character(meta$Disease)
    meta$Disease[meta$Disease == "LUAD"] <- "AD"
    meta$Disease[meta$Disease == "LUSC"] <- "SC"
    meta$Disease[meta$Disease == "LUAD_Normal"] <- "NL(AD)"
    meta$Disease[meta$Disease == "LUSC_Normal"] <- "NL(SC)"
    meta$Disease[meta$Disease == "Pleiomorphic Normal"] <- "NL(UPS)"
    meta$Disease[meta$Disease == "Pleiomorphic"] <- "UPS"
    meta$Disease[meta$Disease == "LULC"] <- "LC"
    meta$Disease[meta$Disease == "LULC_Normal"] <- "NL(LC)"
    set.seed(42)
    meta = meta[sample(1:nrow(meta)), ]
    meta <- meta[order(meta$Stage), c("Stage"), drop = F] # meta$Disease, , "Disease"
    
    
    meta$Cells = rownames(meta)
    num_cells = sort(table(meta$Stage), decreasing = F)
    num_cells = ifelse(num_cells[1] < 20, num_cells[2], num_cells[1])
    meta <- meta %>% group_by(Stage) %>% sample_n(num_cells, replace = T) %>% unique() %>% as.data.frame()
        
        rownames(meta) <- meta$Cells
    
    stage_colors <- c(
        "I"="#65A9A3", 
        "II"="#4A933E",
        "III"="#EC7A21",
        "IV"="#D73F47"
    )
    
    cluster_cols = c(
        as.character(wes_palette("Zissou1")),
        as.character(wes_palette("Royal2")),
        as.character(wes_palette("Royal1")),
        as.character(wes_palette("Darjeeling2")),
        as.character(wes_palette("Darjeeling1"))
    )
    
    patient_colors = cluster_cols[1:length(unique(obj@meta.data$PatientID))]
    names(patient_colors) <- unique(obj@meta.data$PatientID)

    # top annotation
    ta = HeatmapAnnotation(
        Stage = meta$Stage,
        # PatientID = meta$PatientID,
        col = list(
            Stage =  stage_colors
            # PatientID = patient_colors
        ),
        show_legend = T,
        show_annotation_name = T,
        annotation_legend_param = list(
            Stage=list(
                direction = "horizontal", 
                nrow = 1,
                labels_gp = gpar(fontsize = labels_size),
                title_gp = gpar(fontsize = title_size)
            )
            # PatientID=list(
            #     direction = "horizontal",
            #     nrow = 4,
            #     labels_gp = gpar(fontsize = labels_size),
            #     title_gp = gpar(fontsize = title_size)
            # )
        )
    )
    
    
    module_colors = wes_palette("GrandBudapest2", length(unique(res$Clt)), type = "continuous")
    names(module_colors) <- unique(res$Clt)
    
    # left annotation
    if(!is.null(mark_genes)) {
        la = rowAnnotation(
            foo = anno_mark(
                at = which(res$gene %in% mark_genes),
                labels = res$gene[res$gene %in% mark_genes],
                which = "row",
                side = "left"
            ),
            Module = res$Clt,
            col = list(
                Module = module_colors
            ),
            show_legend = F,
            show_annotation_name = F,
            annotation_name_rot = 0,
            annotation_legend_param=list(
                Module=list(
                    direction = "horizontal", 
                    nrow = 1,
                    labels_gp = gpar(fontsize = labels_size),
                    title_gp = gpar(fontsize = title_size)
                )
            )
        )
    } else {
        la = rowAnnotation(
            Module = res$Clt,
            col = list(
                Module = module_colors
            ),
            show_legend = T,
            show_annotation_name = T,
            annotation_name_rot = 0,
            annotation_legend_param=list(
                Module=list(
                    direction = "horizontal", 
                    nrow = 1,
                    labels_gp = gpar(fontsize = labels_size),
                    title_gp = gpar(fontsize = title_size)
                )
            )
        )
    }
    
    h <- Heatmap(
        name = "Expr",
        obj@scale.data[as.character(res$gene), as.character(rownames(meta))],
        cluster_rows = F,
        cluster_columns = F,
        show_column_names = F,
        show_row_names = F,
        col = colorRamp2(c(-2, 0, 2), c("purple", "black", "yellow")),
        top_annotation = ta,
        left_annotation = la,
        border = T,
        column_title_gp = gpar(fontsize=0),
        row_title_gp = gpar(fontsize = 0),
        column_split = meta$Stage,
        row_split = res$Clt,
        heatmap_legend_param = list(
            direction = "horizontal",
            labels_gp = gpar(fontsize = labels_size),
            title_gp = gpar(fontsize = title_size)
        ),
        use_raster = T
    )
    return(h)
}
```


# make scatter piechart
```{R}
make_scatterpie_plot <- function(
    object, 
    color="Disease", 
    alpha=0.8, 
    colors=colors, 
    pt.size = 0.1, 
    guide_ncol=1,
    text_size = 8,
    reduction.use = "tsne",
    r=0.05,
    legend.position="none",
    legend.direction="horizontal",
    legend.breaks=waiver(),
    custom_pos = NULL,
    legend.text = 15
) {
    meta = object@meta.data

    coord = object@dr[[reduction.use]]@cell.embeddings
    colnames(coord) = c("x", "y")

    meta = merge(meta, coord, by = 0)

    scatter = eval(
        parse(
            text=paste0(
                "meta %>% group_by(ident, ", color, ") %>% add_tally() %>% group_by(ident) %>% mutate(x1 = median(x), y1 = median(y), perc = n / sum(n)) %>% dplyr::select(ident, x1, y1, perc, ", color, ") %>% unique()"
            )
        )
    )

    colnames(scatter)[colnames(scatter) == color] = "group"

    scatter = dcast(scatter, ident + x1 + y1 ~ group, value.var="perc")

    scatter[is.na(scatter)] = 0
    scatter$r = min(max(meta$x) - min(meta$x), max(meta$y) - min(meta$y)) * r

    clt_color = gg_color_hue(length(unique(meta$ident)))
    names(clt_color) = 1:length(unique(meta$ident))
    
    if (!is.null(custom_pos)) {
        scatter$x1 = custom_pos[scatter$ident, "x1"]
        scatter$y1 = custom_pos[scatter$ident, "y1"]
    }

    p <- ggplot() + 
        geom_point_rast(
            aes(x=x, y=y, colour=factor(ident)), 
            data=meta, 
            alpha=alpha, 
            size=pt.size
        ) +
        geom_scatterpie(
            aes(
                x=x1,
                y=y1, 
                group=ident,
                r=r
            ),
            data=scatter,
            cols=as.character(unique(meta[, color])),
            color=NA
        ) +
        coord_equal() +
        geom_text(
            aes(x=x1, y=y1 - r - 0.5, label=ident),
            data=scatter,
            size = text_size
        ) +
        theme_bw(base_family = "Arial Unicode MS") +
        theme(
            panel.grid = element_blank(),
            legend.position = legend.position,
            axis.text = element_text(size = 15),
            axis.title = element_text(size = 20),
            legend.text = element_text(size = legend.text),
            # legend.background = element_blank()
            legend.direction=legend.direction,
            legend.spacing.x = unit(0.1, 'cm')
        ) +
        scale_color_manual(values=clt_color, breaks = c()) + 
        scale_fill_manual(values=colors, breaks = legend.breaks) +
        guides(fill = guide_legend(ncol = guide_ncol))

    if (reduction.use == "tsne") {
        p <- p + labs(x="tSNE_1", y="tSNE_2", color="", fill = "")
    } else if (reduction.use == "umap") {
        p <- p + labs(x="UMAP1", y="UMAP2", color="", fill="")
    }

    return(p)
}

```


SFTPB	General
MUC1	General
SFPTA1	General
SFTPC	General
ETV5	General
KIAA1324	AT2_s
CYP4B1	AT2_s
CEACAM6	AT2_s
CTNNBIP1	AT2_s
LRP5	AT2_s
WNT5A	AT2_s
AQP5	AT2_s
CP	AT2_s
GSTA1	AT2_s
TCF7L2	AT2_s
HHIP	AT2
NNMT	AT2
AQP1	AT2
ALPL	AT2
WIF1	AT2
CXCL2	AT2
ACSL4	AT2
CA2	AT2
CDKN1A	AT2
LHFPL3-AS2	AT2


```{r}
genes = c(
    "SFTPB", "MUC1", "SFPTA1", # "SFTPC", "ETV5", "KIAA1324", 
    "AGER", "CAV1", 
    "CEACAM6", "CTNNBIP1", "LRP5", "WNT5A",
    # "HOPX", "HTI-56", "PDPN",
    "AQP5", "CP", "AXIN2", "GSTA1", "TCF7L2", # "HHIP", "WIF1", "LHFPL3-AS2"， "CA2", 
    "NNMT", "AQP1", "ALPL", "CXCL2"
    # "ACSL4", "CDKN1A"
)

wnt = c(
    "Wnt1","Wnt10B", "Wnt10A", "Wnt6","Wnt3", "Wnt9B", "Wnt9A","Wnt3A",
    "Wnt2","Wnt16", "Wnt2B", "Wnt4", "Wnt5A","Wnt5B", "Wnt7A",
    "Wnt7B","Wnt8A", "Wnt8B", "Wnt11", "WntD"
)

wnt <- str_to_upper(wnt)
wnt <- wnt[wnt %in% rownames(obj@raw.data)]
genes = genes[genes %in% rownames(obj@raw.data)]
```


```{R}
markers = read.csv(full.path("11_CNV/each_cells/ATII/LUAD/cluster_markers.csv"), row.names = 1, stringsAsFactors = F)

markers <- markers %>%
    group_by(ident) %>%
    top_n(2, wt = avg_logFC) %>%
    as.data.frame()
```


```{R fig.height=6, fig.width=5}
obj@meta.data$ident1 = as.numeric(obj@meta.data$res.0.03) + 1


generic = c("SFTPB", "MUC1")
at1 = c("AGER", "CAV1")
at2_s = c(
    "CYP4B1", "CEACAM6", "CTNNBIP1", "LRP5", 
    "WNT5A", "AQP5", "CP", "AXIN2", "GSTA1", "TCF7L2"
)
at2 = c("NNMT", "AQP1", "ALPL", "CXCL2")

# genes = c(
#     , "SFPTA1", # "SFTPC", "ETV5", "KIAA1324", 
#     
#     # "HOPX", "HTI-56", "PDPN",
#     , # "HHIP", "WIF1", "LHFPL3-AS2"， "CA2", 
#     
#     # "ACSL4", "CDKN1A"
# )


mark_genes = markers$gene[markers$ident %in% c(1)]
mark_genes = c(mark_genes, "AZGP1")
mark_genes = c(mark_genes, markers$gene[markers$ident %in% 2:4])
mark_genes = c(mark_genes, "EEF1A2")
mark_genes = c(mark_genes, markers$gene[markers$ident == 5], c("KPNA2"))
mark_genes = c(mark_genes, markers$gene[markers$ident > 5])
mark_genes = c(mark_genes, "EGFR")


cols = as.character(wes_palette("Darjeeling1"))

genes = unique(c(mark_genes, generic, at1, at2_s, at2))
a = rep("black", length(genes))

a[genes %in% generic] = cols[4]
a[genes %in% at1] = cols[2]
a[genes %in% at2] = cols[5]
a[genes %in% at2_s] = cols[1]

p <- DotPlot(
    obj, 
    genes.plot = genes, 
    do.return = T, x.lab.rot = T, 
    group.by = "ident1"
) + coord_flip() +
    theme_bw(base_family = "Arial Unicode MS") +
    theme(
        legend.position = "right",
        panel.grid = element_blank(),
        axis.text = element_text(size = 12),
        legend.text = element_text(size = 12),
        legend.title = element_text(size = 15),
        axis.text.y = element_text(colour = rev(a))
    ) +
    labs(x = "", y = "") +
    scale_size_continuous(range = c(0, 5))


p

# ggsave(
#     filename = full.path("11_CNV/each_cells/ATII/LUAD/cluster_stem_dotplot.pdf"),
#     plot = p, width = 5, height = 6, device = cairo_pdf
# )
```


```{r}
meta <- readRDS(full.path("11_CNV/meta.rds"))
obj@meta.data$ident1 = as.numeric(obj@meta.data$res.0.03) + 1
obj@meta.data$Malignant <- as.character(meta[rownames(obj@meta.data), "Malignant_tree"])

temp = obj@meta.data %>%
    group_by(ident1, Malignant) %>%
    add_tally() %>%
    dplyr::select(ident1, Malignant, n) %>%
    unique() %>%
    group_by(ident1) %>%
    mutate(p = n / sum(n)) %>%
    as.data.frame()

ggplot(temp, aes(x=ident1, y=p, fill=Malignant)) +
    geom_bar(stat = "identity")
```

```{r}
temp = obj@meta.data %>%
    group_by(ident1, Malignant) %>%
    add_tally() %>%
    dplyr::select(ident1, Malignant, n) %>%
    unique() %>%
    group_by(Malignant) %>%
    mutate(p = n / sum(n)) %>%
    as.data.frame()

ggplot(temp, aes(x=Malignant, y=p, fill=as.character(ident1))) +
    geom_bar(stat = "identity") +
    labs(x="Malignant subtypes", fill="AT2 clusters", y="") +
    theme_bw() +
    theme(
        axis.text = element_text(size = 15),
        legend.text = element_text(size = 15),
        axis.title = element_text(size = 15)
    )
```


```{r fig.height=4, fig.width=4}
temp = as.data.frame(dm@dm@eigenvectors)
obj@meta.data$ident1 = as.numeric(obj@meta.data$res.0.03) + 1
temp$ident = obj@meta.data[rownames(temp), "ident1"]
temp = temp[!is.na(temp$ident), ]


ggplot(temp, aes(x=DC1, y=DC2, color=as.character(ident))) +
    geom_point_rast(size = 0.1) +
    theme_bw() +
    theme(
        axis.text = element_blank(),
        legend.text = element_text(size = 15),
        legend.title = element_blank(),
        axis.title = element_text(size = 15),
        panel.grid = element_blank(),
        legend.position = c(0.1, 0.2)
    )+
    guides(color = guide_legend(override.aes = list(size = 5)))
```


```{r}
FeaturePlot(obj, features.plot = "AXIN2", cols.use = c("grey", "blue"), reduction.use = "umap")
```


```{r}
meta = readRDS(full.path("11_CNV/meta.rds"))

temp = obj@meta.data
temp$Malignant = meta[rownames(temp), "Malignant_tree"]

temp = temp %>%
    group_by(ident1, Malignant) %>%
    add_tally() %>%
    dplyr::select(ident1, Malignant, n) %>%
    unique() %>%
    group_by(ident1) %>%
    mutate(p = n / sum(n)) %>%
    as.data.frame()


ggplot(temp, aes(x=ident1, y=p, fill=Malignant)) +
    geom_bar(stat = "identity")
```


```{r}
set.seed(42)
temp_meta = obj@meta.data
temp_meta$Cells = rownames(temp_meta)
temp_meta <- temp_meta %>%
    group_by(res.0.03) %>%
    sample_n(min(table(obj@meta.data$res.0.03))) %>%
    as.data.frame()


col_fun = colorRamp2(c(-1, 0, 1), c("purple", "black", "yellow"))

Heatmap(
    obj@scale.data[genes, temp_meta$Cells],
    name = "Expr",
    col = col_fun,
    show_row_names = T,
    show_column_names = F,
    cluster_columns = F,
    cluster_rows = F,
    column_split = temp_meta$res.0.03
)
```



```{r fig.height=4, fig.width=4}
library(ggrastr)
stemness = c("KIAA1324", "CYP4B1", "CTNNBIP1", "LRP5", "WNT5A", "AQP5", "CP", "GSTA1", "TCF7L2")  # "CEACAM6", 

mtx = apply(obj@raw.data, 2, function(col) { col * 10^6 / sum(col) })

mtx <- mtx[stemness, ]
mtx <- na.omit(mtx)

mtx <- melt(as.matrix(mtx))
mtx <- na.omit(mtx)


mtx <- mtx %>%
    group_by(Var2) %>%
    mutate(value = mean(value)) %>%
    dplyr::select(Var2, value) %>%
    unique() %>%
    as.data.frame()

for (i in colnames(obj@dr$umap@cell.embeddings)) {
    mtx[, i] <- obj@dr$umap@cell.embeddings[mtx$Var2, i]
}


plot(density(mtx$value))


mtx$value[mtx$value < 70] = 0

mtx = mtx[order(mtx$value), ]


p <- ggplot(mtx, aes(x=UMAP1, y=UMAP2, color=log2(value + 1))) +
    geom_point_rast(size = 0.1, alpha = 0.5) +
    scale_color_gradient2(low = "white", high = "blue", mid = "grey") +
    theme_bw(base_family = "Arial Unicode MS") +
    theme(
        panel.grid = element_blank(),
        legend.position = c(0.2, 0.25),
        legend.background = element_blank(),
        axis.text = element_text(size = 12),
        axis.title = element_text(size = 15),
        legend.text = element_text(size = 10)
    ) +
    guides(color = guide_colorbar(
        title.position = "bottom"
    ))

ggsave(
    filename = full.path("11_CNV/each_cells/ATII/LUAD/stem_umap.pdf"),
    plot = p, width = 4, height = 4, device = cairo_pdf
)
```


```{r fig.height=4, fig.width=5}
# mtx$val = mtx$value
# mtx$val[mtx$val < 100] = 0
# 
# ggplot(mtx, aes(x=UMAP1, y=UMAP2, color=log2(val + 1))) +
#     geom_point(size = 0.1, alpha = 0.5) +
#     scale_color_gradient2(low = "white", high = "blue", mid = "grey")
```


```{r fig.height=12, fig.width=12}
FeaturePlot(
    obj, 
    c(markers$gene, stemness), 
    reduction.use = "umap", 
    cols.use = c("grey", "blue"), 
    pt.size = 0.1
)
```



### Cluster cell cycle
```{r fig.height=9, fig.width=12}
obj@meta.data$Clt = as.character(as.numeric(obj@meta.data$res.0.3))

VlnPlot(obj, features.plot = wnt, group.by = "Clt", do.sort = T, y.log = T)
```


```{r}
make_violin_plot <- function(obj, gene, group.by = "Clt", scale = T, log=FALSE) {
    meta = obj@meta.data
    meta$value = obj@scale.data[gene, as.character(rownames(meta))]
    
    
    if (log) {
        meta$value = log10(meta$value + 1)
    }
    
    p <- ggviolin(meta, "Clt", "value", fill = "Clt",
       add = "boxplot", add.params = list(fill = "white"))
    
    p
}
```



```{r}
for (i in wnt) {
    p <- make_violin_plot(obj, i, scale = T, log = F) +
        ylim(0, 3) + labs(title = i)
    
    print(p)
}
```


```{r}
go <- read.csv(full.path("../GO_term_summary_20200312_230533.csv"))
go <- str_to_upper(unique(go$Symbol))
go <- intersect(go, rownames(obj@raw.data))


for (i in go) {
    p <- make_feature_plot(obj, i)
    print(p)
}
```




```{r fig.height=6, fig.width=6}
obj@meta.data$ident = as.numeric(obj@meta.data$res.0.03) + 1


p <- make_scatterpie_plot(
    obj, 
    color = "Stage", 
    colors = colors, 
    reduction.use = "umap",
    legend.direction = "horizontal",
    pt.size = 1,
    guide_ncol = 2
)

p = p  + theme(
        text = element_text(size = 10),
        legend.position = c(0.15, 0.1),
        plot.title = element_text(size = 30, face="bold", hjust = 0.5),
        axis.title = element_text(size = 25),
        axis.text = element_text(size = 20),
        legend.text = element_text(size = 20)
    )

p

ggsave(
    filename = full.path("11_CNV/each_cells/ATII/LUAD/ATII_umap_stage.pdf"),
    plot = p, width = 6, height = 6, device = cairo_pdf
)
```



```{r fig.height=6, fig.width=6}


p <- make_scatterpie_plot(
    obj, 
    color = "PatientID", 
    colors = colors, 
    reduction.use = "umap",
    legend.direction = "horizontal",
    pt.size = 1,
    guide_ncol = 3,
    legend.text = 15
)


p = p  + theme(
        text = element_text(size = 10),
        legend.position = c(0.15, 0.1),
        plot.title = element_text(size = 30, face="bold", hjust = 0.5),
        axis.title = element_text(size = 25),
        axis.text = element_text(size = 20),
        legend.spacing = unit(0.05, "cm")
    )


p

ggsave(
    filename = paste0(root.dir, "ATII_umap_patient.pdf"),
    plot = p,
    width = 6,
    height = 6
)
```

```{r fig.height=8, fig.width=15}
pdf(paste0(root.dir, "ATII_combination_bar_plot.pdf"), width = 15, height = 8, onefile = F)
make_combined_barplot(new_obj, group.by = "ident", with_disease = F, ncol = 5)
dev.off()
```

```{r}
find_markers <- function(object, group.by = "Stage", n.cores = 4, min.pct = 0.1) {
    
    temp_group <- as.character(sort(unique(object@meta.data[, group.by])))
    
    library(doMC)
    
    registerDoMC(n.cores)
    res =foreach(i = temp_group, .combine = "rbind") %dopar% {

        groups = rownames(object@meta.data[object@meta.data[, group.by] == i, ])
        
        if(length(groups) >= 3) {
            temp <- FindMarkers(
                object = object, 
                ident.1 = groups,
                logfc.threshold = 0,
                min.pct = min.pct
            )
            
            temp$ident = i
            temp$gene = rownames(temp)
        } 
        
        temp
    }    
    return(res)
}

registerDoMC(5)
obj@meta.data$Clt = as.numeric(as.character(obj@meta.data$res.0.03)) + 1
markers <- find_markers(obj, group.by = "Clt", n.cores = 5)
write.csv(markers, paste0(root.dir, "/cluster_markers.csv"))
```



```{r}
obj@meta.data$ident = as.numeric(obj@meta.data$res.0.03) + 1

prognostic = c("GAPDH","TMSB10","LDHA","TPI1","HSP90AB1","LDHB","RAN","CLIC1","HMGA1","NDRG1","ARF4","PTGES3","ERO1L","NAMPT","ARCN1","RCN3","UGDH","SHMT2","CACYBP","PLIN3")


DotPlot(obj, genes.plot = intersect(prognostic, rownames(obj@raw.data)), group.by = "ident", x.lab.rot = T)
```


```{r}
obj@meta.data$ident = as.numeric(obj@meta.data$res.0.03) + 1

drug <- c("IMPDH2","GAPDH","NUP107","TPI1","COASY")
DotPlot(obj, genes.plot = intersect(drug, rownames(obj@raw.data)), group.by = "ident", x.lab.rot = T)
```

```{r}
obj@meta.data$ident = as.numeric(obj@meta.data$res.0.03) + 1

drug2 <- c("PSMB2","MCM5","MCM2","IARS","CARS","CTPS1","GMPS")
DotPlot(obj, genes.plot = intersect(drug2, rownames(obj@raw.data)), group.by = "ident", x.lab.rot = T)
```

```{r}
mutation <- c("MAP4", "GANAB", "GART", "EGFR", "APEX1", "ALDH18A1", "ATP2A2", "DCXR", "DGAT1", "FLNB", "GALNT3", "NUP188", "PEX14", "PMPCA", "PRKDC", "PRODH", "SLC30A9", "SNRNP200")

obj@meta.data$ident = as.numeric(obj@meta.data$res.0.03) + 1


DotPlot(obj, genes.plot = intersect(mutation, rownames(obj@raw.data)), group.by = "ident", x.lab.rot = T)
```

```{R fig.height=3, fig.width=8}
prognostic = c("TMSB10","LDHA","HSP90AB1","LDHB","RAN","CLIC1","HMGA1","NDRG1","ARF4","PTGES3","ERO1L","NAMPT","ARCN1","RCN3","UGDH","SHMT2","CACYBP","PLIN3", "GAPDH", "TPI1")
drug <- c("IMPDH2","GAPDH","NUP107","TPI1","COASY")
drug2 <- c("PSMB2","MCM5","MCM2","IARS","CARS","CTPS1","GMPS")
mutation <- c("MAP4", "GANAB", "GART", "EGFR", "APEX1", "ALDH18A1", "ATP2A2", "DCXR", "DGAT1", "FLNB", "GALNT3", "NUP188", "PEX14", "PMPCA", "PRKDC", "PRODH", "SLC30A9", "SNRNP200")


cols = as.character(wes_palette("Darjeeling1"))

genes = intersect(
    c(prognostic, drug), 
    rownames(obj@raw.data)
)
a = rep("black", length(genes))

a[genes %in% drug] = cols[2]
a[genes %in% prognostic] = cols[4]
a[genes %in% c("GAPDH", "TPI1")] = "black"

obj@meta.data$ident = as.numeric(obj@meta.data$res.0.03) + 1


p <- DotPlot(
    obj, 
    genes.plot = genes, 
    group.by = "ident", x.lab.rot = T,
    do.return = T
) + 
    theme_bw() +
    theme(
    axis.text.x = element_text(colour = rev(a), angle = 90, hjust = 1, vjust = .5),
    legend.position = "right",
    panel.grid = element_blank(),
    axis.text = element_text(size = 12),
    legend.text = element_text(size = 12),
    legend.title = element_text(size = 15)
) +
    labs(x = "", y = "") +
    scale_size_continuous(range = c(0, 5))

pdf(full.path("11_CNV/each_cells/ATII/LUAD/cluster_proteomic_1.pdf"), width = 8, height = 3)
print(p)
dev.off()

genes = intersect(
    c(drug2, mutation), 
    rownames(obj@raw.data)
)
a = rep("black", length(genes))

a[genes %in% drug2] = cols[5]
a[genes %in% mutation] = cols[1]

obj@meta.data$ident = as.numeric(obj@meta.data$res.0.03) + 1

p <- DotPlot(
    obj, 
    genes.plot = genes, 
    group.by = "ident", x.lab.rot = T,
    do.return = T
)  + 
    theme_bw() +
    theme(
    axis.text.x = element_text(colour = rev(a), angle = 90, hjust = 1, vjust = .5),
    legend.position = "right",
    panel.grid = element_blank(),
    axis.text = element_text(size = 12),
    legend.text = element_text(size = 12),
    legend.title = element_text(size = 15)
) +
    labs(x = "", y = "") +
    scale_size_continuous(range = c(0, 5))

pdf(full.path("11_CNV/each_cells/ATII/LUAD/cluster_proteomic_2.pdf"), width = 8, height = 3)
print(p)
dev.off()
```

### Cluster top markers

```{r fig.height=8, fig.width=6}
obj <- readRDS(full.path("11_CNV/each_cells/ATII/seurat_obj_LUAD.rds"))

markers <- read.csv(
    full.path("11_CNV/each_cells/ATII/cluster_markers_LUAD.csv"), 
    row.names = 1, stringsAsFactors = F
)


set.seed(42)
markers <- markers %>%
    filter(p_val_adj < 0.05 & avg_logFC > 0.5) %>%
    group_by(ident) %>%
    arrange(desc(avg_logFC)) %>%
    as.data.frame()
markers$x = 1:nrow(markers)
    

temp <- markers %>%
    filter(p_val_adj < 0.05 & avg_logFC > 0.25) %>%
    group_by(ident) %>%
    top_n(10, wt=avg_logFC) %>%
    as.data.frame()

obj@meta.data$ident = obj@meta.data$res.0.03

obj@meta.data$Cells = rownames(obj@meta.data)
cells_order <- obj@meta.data %>%
    group_by(ident) %>%
    arrange(ident) %>%
    sample_n(min(table(obj@meta.data$ident))) %>%
    as.data.frame()

# cells_order <- cells_order[sample(1:nrow(cells_order)), ]
# cells_order = cells_order[order(as.numeric(cells_order$ident)), ]

ccls <- as.character(wes_palette("GrandBudapest2", length(unique(cells_order$ident)), type = "continuous"))
names(ccls) <- unique(cells_order$ident)

ta = HeatmapAnnotation(
    Cluster = cells_order$ident,
    col = list(
        Cluster = ccls
    ),
    show_annotation_name = F,
    show_legend = F
)

ha = rowAnnotation(
    foo = anno_mark(
        at = temp$x, 
        labels = temp$gene,
        side = "left"
    )
)


# pdf(paste0(root.dir, "/cluster_markers.pdf"), width = 5, height = 6)
Heatmap(
    obj@scale.data[markers$gene, cells_order$Cells],
    name = "Expr",
    cluster_rows = F,
    cluster_columns = F,
    show_row_names = F,
    show_column_names = F,
    show_heatmap_legend = F,
    top_annotation = ta,
    left_annotation = ha,
    column_split = cells_order$ident,
    row_split = markers$ident,
    row_title = NULL,
    heatmap_legend_param = list(
        labels_gp = gpar(fontsize = 15),
        title_gp = gpar(fontsize = 15)
    ),
    col = colorRamp2(c(-2, 0, 2), c("purple", "black", "yellow"))
)
# dev.off()
```

## Mfuzz
```{r}
markers <- read.csv(full.path("11_CNV/each_cells/ATII/LUAD/stage_markers.csv"), row.names = 1, stringsAsFactors = F)

nms = unique(markers$gene)
ig_genes = c(grep("^IGJ", nms, v=T), 
             grep("^IGH",nms,v=T),
             grep("^IGK", nms, v=T), 
             grep("^IGL", nms, v=T))

bad_genes = unique(c(grep("^MT-", nms, v=T), grep("^MTMR", nms, v=T), grep("^MTND", nms, v=T),"NEAT1","TMSB4X", "TMSB10", ig_genes))



markers <- markers %>% 
    filter(avg_logFC > 0.5 & p_val_adj < 0.05 & !gene %in% bad_genes)


expr = ExpressionSet(
    as.matrix(obj@scale.data[unique(markers$gene),])
)
    
m = mestimate(expr)

temp = Dmin(expr, m=m, crange=seq(2,10,1), repeats=3, visu=FALSE)
print(temp)
if (sum(!is.na(temp)) == 0) {
    c = length(unique(obj@meta.data$Stage))
} else {
    new_temp = temp[!is.na(temp)]
    
    # using kneed to selected best cluster
    kee = import("kneed")

    c = kee$KneeLocator(
        1:length(new_temp),
        new_temp, 
        curve='convex', 
        direction='decreasing'
    )
    c = c$knee
    
    c = which(temp == new_temp[c])
}

cl <- mfuzz(expr, c=3, m=m)


# extract results
res = as.data.frame(cl$cluster)
colnames(res) <- "Clt"
res$gene = rownames(res)

res = res[order(res$Clt), ]


res$Clt <- sapply(res$Clt, function(x) {
    temp = c(
        "1"="M.I",
        "2"="M.II",
        "3"="M.IV"
    )
    
    return(temp[as.character(x)])
})

write.csv(res, full.path("11_CNV/each_cells/ATII/LUAD/mfuzz.csv"))
```

```{r fig.height=6, fig.width=8}
res = read.csv(full.path("11_CNV/each_cells/ATII/LUAD/mfuzz.csv"), row.names = 1, stringsAsFactors = F)

set.seed(42)
temp = res$gene[sample(1:nrow(res))]

labels_size = 12
title_size = 15

pdf(paste0(root.dir, "11_CNV/each_cells/ATII/LUAD/ATII_mfuzz.pdf"), height = 6, width = 8)
draw(
    make_module_heatmap(obj, res, mark_genes = c(head(temp, n = 10), "PIGR", "AZGP1", "BTG2", "EGR1", "AQP5", "SPINK", "CP")), 
    column_title = "",
    column_title_gp = gpar(fontsize = 20),
    merge_legend = TRUE, 
    heatmap_legend_side = "bottom", 
    annotation_legend_side = "bottom"
)
dev.off()
```


```{R}
library(TCGAbiolinks)

lusc_cli <- readRDS("LungCancer10x/TCGA/LUAD_clinical.rds")
tcga <- read.csv(
    "LungCancer10x/TCGA/Lung.csv", 
    row.names = 1, 
    stringsAsFactors = F
)

tcga <- tcga[, str_detect(colnames(tcga), "LUAD")]

for (i in unique(temp$ident)) {
    genes = as.data.frame(temp)[temp$ident == i, "gene"]
    
    expr <- summary(apply(tcga[intersect(genes, rownames(tcga)), ], 1, mean))

    samples <- colnames(tcga)[apply(tcga[intersect(genes, rownames(tcga)), ], 2, function(x) {
        mean(x) > expr[5]
    })]
    
    samples <- sapply(samples, function(x) {
        temp = str_split(x, "\\.")[[1]]
        paste(temp[1], temp[2], temp[3], sep = "-")
    })
    
    lusc_cli$ident = ifelse(lusc_cli$submitter_id %in% samples, "Up", "Down")
    
    TCGAanalyze_survival(
        lusc_cli,
        "ident",
        main = paste("Cluster", i),
        height = 6, 
        width=8,
        risk.table = F,
        dpi = 600,
        conf.int = F,
        cumevents = F,
        cumcensor = F,
        ggtheme = theme_minimal() + theme(
            text = element_text(size = 15),
            axis.text = element_text(size = 15),
            axis.title = element_text(size = 18),
            legend.text = element_text(size = 15),
            legend.title = element_text(size = 18),
            title = element_text(size = 20)
        ),
        filename = paste("ATI_", i, ".pdf", sep = "")
    )
}
```

### cluter annotation
```{R}
cerno <- read.xlsx(here("clusters/tmod.xlsx"), sheet = 1)

cerno <- cerno %>%
    filter(adj.P.Val < 0.05) %>%
    group_by(ident) %>%
    top_n(10, wt = AUC)

cerno <- reshape2::dcast(cerno, Title~ident, value.var = "AUC", fun.aggregate = mean, fill = 0)

rownames(cerno) <- cerno$Title
cerno <- cerno[, colnames(cerno) != "Title"]
```


```{R fig.height=12, fig.width=12}

col_fun <- colorRamp2(c(0.9, 1), c("white", "red"))

h <- Heatmap(
    as.matrix(cerno),
    name = "AUC",
    col = col_fun,
    column_names_rot = 0,
    column_names_centered = T,
    column_names_gp = gpar(fontsize = 20),
    row_names_gp = gpar(fontsize = 15),
    heatmap_legend_param = list(
        direction = "horizontal",
        labels_gp = gpar(fontsize = labels_size),
        title_gp = gpar(fontsize = title_size),
        title_position = "lefttop"
    ),
    border = T,
    width = unit(6, "cm"),
    rect_gp = gpar(col = "grey75", lwd = 1)
)

pdf(here("clusters/heatmap_cerno.pdf"), width = 12, height = 12)
draw(
    h, 
    heatmap_legend_side = "bottom"
)
dev.off()
```


### Make a cluster plot of all LUSC Basal cells
```{r}
meta <- read.csv(paste0(root.dir, "../../02_rds/meta_after_singleR.csv"), row.names = 1, stringsAsFactors = F)

nm_meta = read.xlsx(paste0(root.dir, "../../02_rds/NM_meta.xlsx"))
rownames(nm_meta) <- nm_meta$SampleID

meta$Disease[meta$SampleID %in% nm_meta$SampleID] = nm_meta[meta$SampleID[meta$SampleID %in% nm_meta$SampleID], "Disease"]
meta$Stage[meta$SampleID %in% nm_meta$SampleID] = nm_meta[meta$SampleID[meta$SampleID %in% nm_meta$SampleID], "Stage"]

meta <- meta[as.character(meta$cell_short) == "ATII", ]
meta <- meta[as.character(meta$Disease) == "LUAD", ]

expr <- readRDS(paste0(root.dir, "../../02_rds/all_cell_expr.rds"))
```

```{r}
all_obj <- CreateSeuratObject(
    expr[, rownames(meta)],
    meta.data = meta
)

mito.genes <- grep(pattern = "^MT-", x = rownames(x = all_obj@raw.data), value = TRUE)
percent.mito <- Matrix::colSums(all_obj@raw.data[mito.genes, ])/Matrix::colSums(all_obj@raw.data)

# AddMetaData adds columns to object@meta.data, and is a great place to
# stash QC stats
all_obj <- AddMetaData(object =all_obj, metadata = percent.mito, col.name = "percent.mito")


### 3. do PCA and harmony

pcs = min(100, min(nrow(all_obj@raw.data), ncol(all_obj@raw.data)) - 1)

all_obj <- NormalizeData(object = all_obj, normalization.method = "LogNormalize", scale.factor = 10000)
all_obj <- FindVariableGenes(object = all_obj, mean.function = ExpMean, dispersion.function = LogVMR, x.low.cutoff = 0.0125, x.high.cutoff = 3, y.cutoff = 0.5)
all_obj <- ScaleData(object = all_obj, vars.to.regress = c("nUMI", "percent.mito"))
all_obj <- RunPCA(object = all_obj, pc.genes = all_obj@var.genes, do.print = FALSE, pcs.compute = pcs)

all_obj <- RunHarmony(all_obj, c("batch", "PatientID"))


#### 4. make dotplot of SD of PCAs and Harmonys

temp <- as.data.frame(apply(all_obj@dr$harmony@cell.embeddings, 2, sd))
colnames(temp) <- c("sd")
temp$Harmony = 1:nrow(temp)

kee = import("kneed")
n.pcs = kee$KneeLocator(
    1:ncol(all_obj@dr$pca@cell.embeddings),
    apply(all_obj@dr$pca@cell.embeddings, 2, sd), 
    curve='convex', 
    direction='decreasing'
)
# n.pcs = n.pcs$knee
n.pcs = 10


#### 5. tSNE and UMAP
all_obj <- RunTSNE(
    object = all_obj, 
    dims.use = 1:n.pcs, 
    do.fast = TRUE, 
    reduction.use = "harmony", 
    reduction.name = "tsne",
    perplexity=min(30, floor(pcs / 3) + 1)
)
all_obj <- RunUMAP(
    object = all_obj, 
    dims.use = 1:n.pcs, 
    do.fast = TRUE, 
    reduction.use = "harmony",
    n_neighbors = min(30, pcs)
)
all_obj <- FindClusters(object = all_obj, reduction.type = "harmony", dims.use = 1:10, resolution = c(0.1, 0.2, 0.3, 0.4, 0.5), print.output = 0, save.SNN = FALSE, force.recalc = TRUE)


saveRDS(all_obj, paste0(root.dir, "total_atii.rds"))
```


### Var genes
```{r fig.height=8, fig.width=12}
ta = HeatmapAnnotation(
    Patient = order(all_obj@meta.data$PatientID)
)

Heatmap(
    name = "Expr",
    all_obj@scale.data[all_obj@var.genes[1:500], rownames(all_obj@meta.data)[order(all_obj@meta.data$PatientID)]],
    cluster_rows = F,
    cluster_columns = F,
    show_row_names = F,
    show_column_names = F,
    top_annotation = ta,
    col = colorRamp2(c(-2, 0, 2), c("purple", "black", "yellow"))
)
```



```{r eval=FALSE, include=FALSE}
all_obj <- readRDS(paste0(root.dir, "total_atii.rds"))
```


```{r}
all_obj@meta.data$ident1 <- as.numeric(all_obj@meta.data$res.0.1)
all_obj@meta.data$ident <- all_obj@meta.data$ident1

p <- DimPlot(
    all_obj, 
    reduction.use = "umap", 
    pt.size = 0.1, 
    do.label = T, 
    no.legend = T,
    do.return = T,
    label.size = 10,
    group.by = "ident1"
)

p <- p + 
    theme_bw() +
    theme(
        text = element_text(size = 10),
        legend.position = "none",
        plot.title = element_text(size = 30, face="bold", hjust = 0.5),
        axis.title = element_text(size = 25),
        axis.text = element_text(size = 20),
        aspect.ratio = 1
    ) +
    labs(title = "ATII (LUAD)")

p

# ggsave(
#     filename = paste0(root.dir, "ATII_all_umap_cluster.pdf"),
#     plot = p,
#     width = 6,
#     height = 6
# )
```


```{r fig.height=6, fig.width=6}
p <- make_scatterpie_plot(
    all_obj, 
    color = "Stage", 
    colors = colors, 
    reduction.use = "umap",
    legend.position = c(0.8, 0.08),
    guide_ncol = 3
) + theme(aspect.ratio = 1)

p

# ggsave(
#     filename = paste0(root.dir, "ATII_all_umap_stage.pdf"),
#     plot = p,
#     width = 6,
#     height = 6
# )


cluster_cols = c(
    as.character(wes_palette("Zissou1")),
    as.character(wes_palette("Royal2")),
    as.character(wes_palette("Royal1")),
    as.character(wes_palette("Darjeeling2")),
    as.character(wes_palette("Darjeeling1"))
)
patient_colors = cluster_cols[1:length(unique(as.character(all_obj@meta.data$PatientID)))]
names(patient_colors) <- as.character(unique(all_obj@meta.data$PatientID))
colors = c(stage_colors, disease_colors, tissue_colors, patient_colors)


p <- make_scatterpie_plot(
    all_obj, 
    color = "PatientID", 
    colors = colors, 
    reduction.use = "umap",
    legend.position = c(0.85, 0.8),
    guide_ncol=2,
    legend.text = 8
) + theme(aspect.ratio = 1)

p

# ggsave(
#     filename = paste0(root.dir, "ATII_all_umap_patient.pdf"),
#     plot = p,
#     width = 6,
#     height = 6
# )
```


```{r fig.height=8, fig.width=15}
# pdf(paste0(root.dir, "ATII_all_combination_bar_plot.pdf"), width = 15, height = 8, onefile = F)
make_combined_barplot(all_obj, group.by = "ident", with_disease = F, ncol = 5)
# dev.off()
```



```{r fig.height=3, fig.width=4}
calculate_cluster_perc <- function(obj, seed.range = 1:50) {
    meta = obj@meta.data

    num = table(meta$Stage)

    sf = num / min(num)
    
    # sf = table(meta$ident)
    # sf = sf / min(sf)
    
    temp <- meta %>%
        group_by(ident, Stage) %>%
        add_tally() %>%
        dplyr::select(ident, Stage, n) %>%
        unique() %>%
        as.data.frame()
    
    temp$freq <- apply(temp, 1, function(row) {
        as.numeric(row[3]) / sf[[row[2]]] # / num[[row[2]]]
    })
    
    temp <- temp %>% group_by(ident) %>%
        mutate(freq = freq / sum(freq)) %>%
        as.data.frame()
    
    temp
}

atii_perc <- calculate_cluster_perc(all_obj)

p <- ggplot(atii_perc, aes(x=factor(ident), y = freq, color = Stage)) + 
    geom_point() +
    geom_line(aes(x=ident, y = freq, color = Stage)) +
    scale_color_manual(values = c(
        "I"="#65A9A3", 
        "II"="#4A933E",
        "III"="#EC7A21",
        "IV"="#D73F47"
    )) +
    labs(x = "", y = "", color = "") +
    theme_bw() +
    theme(
        axis.text = element_text(size = 15),
        legend.text = element_text(size = 15),
        legend.position = c(0.1, 0.80),
        legend.background = element_blank()
    )
p


ggsave(
    filename = paste0(root.dir, "ATII_all_stage_perc.pdf"),
    plot = p,
    width = 4,
    height = 3
)
```


```{r}
atii_markers <- find_markers(all_obj, group.by = "ident", n.cores = 5, min.pct = 0.1)

write.csv(atii_markers, paste0(root.dir, "ATII_all_cells_cluster_markers.csv"))
```

```{R fig.height=10, fig.width=8}

atii_markers <- atii_markers[order(atii_markers$avg_logFC, decreasing = T), ]

temp <- atii_markers %>%
    filter(p_val_adj < 0.05) %>%
    group_by(ident) %>%
    top_n(10, wt=avg_logFC)


cells_order = rownames(all_obj@meta.data)[order(as.character(all_obj@meta.data$ident))]

ccls <- cluster_cols[1:length(unique(all_obj@meta.data$ident))]
names(ccls) <- unique(all_obj@meta.data$ident)

ta = HeatmapAnnotation(
    Cluster = as.character(all_obj@meta.data$ident)[order(as.character(all_obj@meta.data$ident))],
    col = list(
        Cluster = ccls
    ),
    show_annotation_name = F,
    show_legend = F
)


# pdf(paste0(root.dir, "ATII_cluster_markers_heatmap.pdf"), width = 6, height = 8)
Heatmap(
    name = "Expr",
    all_obj@scale.data[temp$gene, cells_order],
    cluster_rows = F,
    cluster_columns = F,
    show_row_names = T,
    show_column_names = F,
    top_annotation = ta,
    column_split = as.character(all_obj@meta.data$ident[order(as.character(all_obj@meta.data$ident))]),
    row_split = as.character(temp$ident),
    heatmap_legend_param = list(
        labels_gp = gpar(fontsize = 15),
        title_gp = gpar(fontsize = 15)
    ),
    col = colorRamp2(c(-2, 0, 2), c("purple", "black", "yellow"))
)
# dev.off()
```

## make violin plot
```{r fig.height=30, fig.width=16}
VlnPlot(all_obj, features.plot = sort(temp$gene), group.by = "res.0.1", point.size.use = 0)
```

### make final cluste vln plot
```{r fig.height=8, fig.width=4}
target_genes = c(
    "CYP2B7P", "MUC1",
    "CKS1B", "H2AFZ", 
    "LYZ", "SRGN",
    "FABP5", "SFTPC", "SFTPD"
)

temp_expr <- melt(as.matrix(all_obj@scale.data[target_genes, ]))
temp_expr$ident <- all_obj@meta.data[temp_expr$Var2, "res.0.1"]

library(ggpubr)
p <- ggviolin(
    data = temp_expr,
    x = "ident",
    y = "value",
    fill = "ident",
    add = "boxplot", 
    add.params = list(fill = "white")
) +
    facet_grid(Var1~., scales = "free") +
    labs(x = "", y = "") +
    theme_bw() +
    theme(
        legend.position = "none",
        axis.text.y = element_text(size = 12),
        axis.text.x = element_text(size = 15),
        strip.text = element_text(size = 12)
    )

ggsave(
    filename = paste0(root.dir, "ATII_cluster_markers_vln.pdf"),
    plot = p,
    width = 4,
    height = 8
)
```



```{r fig.height=6, fig.width=10}
res_markers <- readRDS("Lung_cancer_10x/ATII/new_markers_multiple_seed.rds")

temp_markers <- res_markers[str_detect(res_markers$ident, "LUAD-"), c(2, 1)]
colnames(temp_markers) <- c("gene", "Clt")


temp_meta <- expr@meta.data[expr@meta.data$cell_name == "Alveolar II" & expr@meta.data$Disease == "LUAD", ]

temp_obj1 <- CreateSeuratObject(
    expr@raw.data[, rownames(temp_meta)],
    meta.data = temp_meta
)

temp_obj1@scale.data = expr@scale.data[, rownames(temp_meta)]


make_module_heatmap(temp_obj1, temp_markers, col_fun = colorRamp2(c(-1, 0, 1), c("purple", "black", "yellow")))
```

### Corrlation
```{R fig.height=8, fig.width=8}
stage_cor <- cor(t(obj@scale.data[res$gene, ]))

module_colors = wes_palette("GrandBudapest2", length(unique(res$Clt)), type = "continuous")
names(module_colors) <- unique(res$Clt)


ba <- HeatmapAnnotation(
    Module=res$Clt,
    col = list(
        Module=module_colors
    )
)


ra <- rowAnnotation(
    Module=res$Clt,
    col = list(
        Module=module_colors
    ),
    show_legend = F,
    show_annotation_name = F
)


pdf(here("mfuzz/module_gene_cor.pdf"), width = 8, height = 6)
p <- Heatmap(
    stage_cor[res$gene, res$gene],
    name = "Corr",
    show_row_names = F,
    show_column_names = F,
    right_annotation = ra,
    bottom_annotation = ba
)
draw(p, merge_legend=T)
dev.off()
# ggcorrplot::ggcorrplot(stage_cor)
```

```{r}

stage_cor <- cor(obj@scale.data[res$gene, ])

module_colors = wes_palette("GrandBudapest2", length(unique(res$Clt)), type = "continuous")
names(module_colors) <- unique(res$Clt)

meta <- obj@meta.data
meta <- meta[order(meta$Stage, meta$PatientID), ]

ba <- HeatmapAnnotation(
    Stage=meta$Stage,
    Patient=meta$PatientID,
    col = list(
        Stage=stage_colors,
        Patient=patient_colors
    )
)


ra <- rowAnnotation(
    Stage=meta$Stage,
    Patient=meta$PatientID,
    col = list(
        Stage=stage_colors,
        Patient=patient_colors
    ),
    show_legend = F,
    show_annotation_name = F
)


h <- Heatmap(
    stage_cor[rownames(meta), rownames(meta)],
    name = "Corr",
    show_row_names = F,
    show_column_names = F,
    right_annotation = ra,
    bottom_annotation = ba,
    cluster_rows = F,
    cluster_columns = F
)

pdf(here("mfuzz/module_cell_cor.pdf"), width = 8, height = 6)
draw(h, merge_legend=T)
dev.off()
```

### Radar plots on bulk data
```{r}
bulk_meta <- read.xlsx(here("../../../09_bulk/RNATAC50commonSample-1119.xlsx"))
bulk_meta <- bulk_meta[!is.na(bulk_meta$SampleID), ]
rownames(bulk_meta) <- bulk_meta$SampleID
```

### MuSiC
```{r}
bulk <- read.xlsx(here("../../../09_bulk/RNA_seq/RSEM.xlsx"), rowNames = T)
bulk <- bulk[, str_detect(colnames(bulk), "LUSC")]

estimate <- readRDS(here("../../../09_bulk/RNA_seq/MuSic/bulk.rds"))
estimate <- estimate$Est.prop.weighted
estimate <- as.matrix(t(estimate))

for (i in colnames(bulk)) {
    if (estimate["Basal", i] > 0) {
        bulk[, i] <- bulk[, i] / estimate["ATII", i]
    } else {
        bulk[, i] <- 0
    }
    
}

bulk <- scale(bulk)
bulk <- melt(as.matrix(bulk))
bulk <- na.omit(bulk)
```

```{r}
gene_pos <- read.table(here("../../../09_bulk/RNA_seq/gene_pos.txt"))
rownames(gene_pos) <- gene_pos$V5

bulk$Var1 <- gene_pos[bulk$Var1, "V6"]

bulk <- merge(bulk, res, by.x = "Var1", by.y = "gene")
bulk$Stage <- bulk_meta[bulk$Var2, "Stage"]
bulk$Stage <- str_replace_all(bulk$Stage, "[^IV]", "")
bulk <- bulk[bulk$Stage != "", ]
```

```{r}
temp <- dcast(bulk, Var1~Var2, fun.aggregate = mean, value.var = "value")
rownames(temp) <- temp$Var1
temp <- temp[, colnames(temp) != "Var1"]

rownames(res) <- res$gene

temp1 = as.matrix(temp[rowSums(temp) > 0, ])
col_fun <- colorRamp2(c(-1, 0, 1), c("blue", "white", "red"))

pdf(here("mfuzz/mfuzz_bulk_MuSic.pdf"), width = 7, height = 15) 
Heatmap(
    temp1,
    name = "Expr",
    cluster_columns = T,
    cluster_rows = T,
    show_row_names = F,
    show_column_names = F,
    row_split = res[rownames(temp1), "Clt"],
    column_split = str_replace_all(bulk_meta[colnames(temp1), "Stage"], "[^IV]", ""),
    col = col_fun,
    border = T
)
dev.off()
```

### Radar plots of MuSiC results
```{r}
temp <- bulk %>%
    group_by(Clt, Stage, Var2, Var1) %>%
    mutate(value = mean(value)) %>%
    unique() %>%
    group_by(Clt, Stage, Var2) %>%
    mutate(value = mean(value)) %>%
    dplyr::select(Var2, value, Clt, Stage) %>%
    unique()
```


```{r}
for(i in unique(temp$Stage)) {
    temp1 <- dcast(temp[temp$Stage == i, ], Var2~Clt, fun.aggregate = mean, value.var = "value")

    p <- ggradar(
        temp1,
        centre.y = -0.1,
        values.radar = round(seq(
            from=-0.1,
            to=0.1,
            length.out = 3
        ), 2),
        grid.max = 0.1,
        grid.label.size = 8,
        legend.position = "bottom",
        legend.title = "",
        plot.title = i,
        axis.label.size = 5,
        group.line.width = 0.5,
        group.point.size = 4,
        axis.label.offset = 1.15,
        base.size = 10
    )
    p = p + theme(
        legend.direction="horizontal",
        legend.spacing.x = unit(0.2, 'cm'),
        legend.position = c(0.5, 0.01),
        legend.background = element_blank(),
        legend.text = element_text(size = 20),
        plot.title = element_text(size = 25, hjust = 0.5)
    ) +
    guides(
        color = guide_legend(override.aes = list(ncol = 3))
    )

    ggsave(
        filename = here(paste0("mfuzz/mfuzz_radar_MuSiC_", i, ".pdf")),
        plot = p,
        width = 6,
        height = 6
    )
}
```

### Stage module radar plot
```{R}
format_radar_expr_mat <- function(
  obj, 
  module,
  legend.position="bottom",
  plot.title="",
  legend.title = "",
  axis.label.size = 8,
  batch=F,
  Stage = NULL
) {
    meta = as.data.frame(obj@meta.data)
    
    if (!is.null(Stage)) {
        meta = meta[meta$Stage == Stage, ]
    }
    
    if (batch) {
        meta = meta[meta$Batch == 3, ]
        meta$Group = meta$PatientID
    } else {
        meta = meta[meta$Batch != 3, ]
        meta$Group = meta$PatientID # paste0(meta$PatientID, "(", meta$Stage, ")")
    }
    

    cells = meta[, "Group", drop = F]
    cells$Cells = rownames(cells)
    
    mat = obj@scale.data[as.character(module$gene), ]
    
    mat = melt(as.matrix(mat))
    mat = merge(cells, mat, by.x = "Cells", by.y = "Var2")
    
    module$Clt = as.character(module$Clt)
    # module$Clt = sapply(module$Clt, function(x) { paste0("M.", x) })
    
    module = merge(module, mat, by.x = "gene", by.y = "Var1", all.x = T)
    
    
    module = module %>% group_by(Clt, Group, gene) %>%
        mutate(m = mean(value)) %>%
        dplyr::select(Clt, Group, m) %>%
        unique() %>%
        group_by(Clt, Group) %>%
        mutate(value = mean(m)) %>%
        dplyr::select(Clt, Group, value) %>%
        unique()
    
    # return(module)
    module = as.data.frame(module)
    
    # return(module)
    #print(module)
    # 
    # module
    p <- ggradar(
        dcast(module, Group~Clt),
        centre.y = floor(min(module$value)),
        values.radar = round(seq(from=floor(min(module$value)), to=max(module$value), length.out = 3), 2),
        grid.max = ceiling(max(module$value)),
        grid.label.size = 8,
        legend.position = legend.position,
        legend.title = legend.title,
        plot.title = plot.title,
        axis.label.size = axis.label.size,
        group.line.width = 0.5,
        group.point.size = 4,
        axis.label.offset = 1.15,
        base.size = 10
    )
    p = p + theme(
        legend.direction="horizontal",
        legend.spacing.x = unit(0.2, 'cm'),
        legend.position = c(0.5, 0.01),
        legend.background = element_blank(),
        legend.text = element_text(size = 20),
        plot.title = element_text(size = 25, hjust = 0.5)
    ) +
        guides(
            color = guide_legend(override.aes = list(ncol = 3))
        )
    return(p)
}

```

```{R fig.height=6, fig.width=6}

for (i in unique(obj@meta.data$Stage)) {
    p = format_radar_expr_mat(
        obj, 
        res,
        legend.position="bottom",
        # plot.title=basename(dirname(dirname(i))),
        plot.title = paste0("ATII ", "(", i, ")"),
        axis.label.size = 10,
        Stage=i
    )
    
    ggsave(
        filename = here(paste0("mfuzz/mfuzz_radar_", i, ".pdf")),
        plot = p,
        width = 6,
        height = 6
    )
}


```

```{r}
# obj@meta.data$PatientID <- as.character(obj@meta.data$PatientID)
# obj@meta.data$Disease <- as.character(obj@meta.data$Disease)
# obj@meta.data$PatientID[obj@meta.data$Batch == 3] <- meta[as.character(obj@meta.data$SampleID[obj@meta.data$Batch == 3]), "PatientID"]
# obj@meta.data$Disease[obj@meta.data$Batch == 3] <- meta[as.character(obj@meta.data$Disease[obj@meta.data$Batch == 3]), "Disease"]
# 
# 
# cells = rownames(obj@meta.data)[obj@meta.data$Batch == 3]
# 
# temp_obj <- CreateSeuratObject(
#     obj@raw.data[, cells],
#     meta.data = obj@meta.data[cells, ]
# )
# 
# temp_obj@scale.data = obj@scale.data[, cells]
```


## format meta info

```{R}
expr <- readRDS(here("../../../02_rds/seurat_obj_nm.rds"))
```


```{r}
meta <- read.xlsx(here("../../../02_rds/NM_meta.xlsx"))

rownames(meta) <- meta$SampleID
meta$PatientID <- paste0("DL", meta$PatientID)

expr@meta.data$PatientID <- as.character(expr@meta.data$PatientID)
expr@meta.data$Disease <- as.character(expr@meta.data$Disease)
expr@meta.data$PatientID[expr@meta.data$Batch == 3] <- meta[as.character(expr@meta.data$SampleID[expr@meta.data$Batch == 3]), "PatientID"]
expr@meta.data$Disease[expr@meta.data$Batch == 3] <- meta[as.character(expr@meta.data$SampleID[expr@meta.data$Batch == 3]), "Disease"]
expr@meta.data$Stage[expr@meta.data$Batch == 3] <- meta[as.character(expr@meta.data$SampleID[expr@meta.data$Batch == 3]), "Stage"]
```


```{r fig.height=6, fig.width=6}
temp_meta <- expr@meta.data[expr@meta.data$cell_name == "Alveolar II" & expr@meta.data$Disease %in% c("LUAD"), ]

temp_obj <- CreateSeuratObject(
    expr@raw.data[, rownames(temp_meta)],
    meta.data = temp_meta
)

temp_obj@scale.data = expr@scale.data[, rownames(temp_meta)]

temp_obj@meta.data$PatientID <- paste(temp_obj@meta.data$PatientID, temp_obj@meta.data$Stage, sep = "-")

p <- format_radar_expr_mat(
    temp_obj, 
    res,
    legend.position="bottom",
    # plot.title=basename(dirname(dirname(i))),
    plot.title = "ATII (DL2018)",
    axis.label.size = 10,
    batch = T
)

p

ggsave(
    filename = here(paste0("mfuzz/mfuzz_radar_DL.pdf")),
    plot = p,
    width = 6,
    height = 6
)
```



### make violin plot

DEGs between clusters
```{r fig.height=6, fig.width=6}
gene_drug_luad <- c("ALKBH3", "ALOX5AP", "BRAF", "CASP8", "CCDC130", "EGFR", "EIF2S1", "ERP27", "G6PD", "INSR", "KRAS", "MET", "PSMB4", "RBPJ", "ROS1", "RPS6KA1", "SYT12", "TNFSF9")

DotPlot(
    all_obj, 
    genes.plot = gene_drug_luad, 
    x.lab.rot = T, 
    group.by = "res.0.8", 
    do.return = T
) +
    coord_flip()
```


```{r}
format_expr_by_mean_for_heatmap <- function(object, group.by, genes.use=NULL) {
    if (is.null(genes.use)) {
        genes.use = rownames(object@scale.data)
    }
    
    expr <- reshape2::melt(object@scale.data[genes.use, ])
    expr$group = object@meta.data[expr$Var2, group.by]
    
    expr <- expr %>% 
        group_by(Var1, group) %>%
        mutate(value = mean(value)) %>%
        dplyr::select(Var1, group, value) %>%
        unique()
    
    expr <- reshape2::dcast(expr, Var1~group, fun.aggregate = mean)
    
    rownames(expr) <- expr$Var1

    return(expr[, colnames(expr) != "Var1"])
}


expr <- format_expr_by_mean_for_heatmap(obj, group.by = "Disease", genes.use = gene_drug_luad)
colnames(expr) <- sapply(colnames(expr), function(x) {
    renames <- c(
        "LUAD"="LUAD",
        "LUSC"="LUSC",
        "LUAD_Normal"="Norm.(LUAD)",
        "LUSC_Normal"="Norm.(LUSC)"
    )
    
    renames[[x]]
})
```


```{r fig.height=6, fig.width=8}
expr <- format_expr_by_mean_for_heatmap(all_obj, group.by = "ident", genes.use = gene_drug_luad)
pdf(paste0(root.dir, "ATII_gene_drug_heatmap_clusters.pdf"), width = 8, height = 6)
Heatmap(
    as.matrix(expr),
    heatmap_legend_param = list(
        title = "Mean expr.",
        labels_gp = gpar(fontsize = labels_size),
        title_gp = gpar(fontsize = title_size)
    ),
    column_names_rot = 30,
    column_names_gp = gpar(fontsize = 20),
    row_names_gp = gpar(fontsize = 15),
    col = colorRamp2(c(-0.5, 0, 0.5), c("blue", "white", "red"))
)
dev.off()
```

```{r fig.height=6, fig.width=8}
expr_clt <- format_expr_by_mean_for_heatmap(obj, group.by = "res.0.8", genes.use = gene_drug_luad)

pdf(here("gene_drug_vln/heatmap_clusters.pdf"), width = 8, height = 6)
Heatmap(
    as.matrix(expr_clt),
    heatmap_legend_param = list(
        title = "Mean expr.",
        # legend_height = unit(3, "cm"),
        # legend_width = unit(10, "cm"),
        labels_gp = gpar(fontsize = labels_size),
        title_gp = gpar(fontsize = title_size)
    ),
    column_names_rot = 0,
    column_names_gp = gpar(fontsize = 20),
    row_names_gp = gpar(fontsize = 15)
)
dev.off()
```


##### First Round
```{r}
read_protein <- function(path) {
    
    wb <- loadWorkbook(path)
    sheets <- wb$sheet_names
    
    res = NULL
    data <- read.xlsx(wb, sheet = which(sheets == "SNR532"))
    data <- data[, !colnames(data) %in% c("Block", "Column", "Row")]
    # data[, 3:ncol(data)] <- scale(data[, 3:ncol(data)])
    data <- reshape2::melt(data, id.vars = c("Name", "ID"))
    data$source = "SNR532"
    res = rbind(data, res)
    
    data <- read.xlsx(wb, sheet = which(sheets == "SNR635"))
    data <- data[, !colnames(data) %in% c("Block", "Column", "Row")]
    # data[, 3:ncol(data)] <- scale(data[, 3:ncol(data)])
    data <- reshape2::melt(data, id.vars = c("Name", "ID"))
    data$source = "SNR635"
    
    res = rbind(res, data)
    
    res$variable = as.character(res$variable)
    res$variable[str_detect(res$variable, "^LA")] <- "LUAD"
    res$variable[str_detect(res$variable, "^LS")] <- "LUSC"
    res$variable[str_detect(res$variable, "^NC")] <- "Normal"
    res <- res[res$variable %in% c("LUAD", "LUSC", "Normal"), ]
    
    return(res)
}


data1 <- read_protein(here("../../../10_protein/1_round.xlsx"))
data2 <- read_protein(here("../../../10_protein/2_round.xlsx"))
```



### protein heatmap
```{r fig.height=6, fig.width=6}
temp = data1[data1$Name %in% gene_drug_luad & data1$source == "SNR532", ]

temp = reshape2::dcast(temp, Name~variable, value.var = "value", fun.aggregate = mean)
rownames(temp) <- temp$Name
temp <- temp[, colnames(temp) != "Name"]

temp <- t(scale(t(temp)))

col_fun <- colorRamp2(c(-1, 0, 1), c("blue", "white", "red"))

pdf(here("gene_drug_vln/heatmap_SNR532_IgG.pdf"), width = 6, height = 6) 
Heatmap(
    temp,
    # t(scale(t(temp))),
    col = col_fun,
    column_names_rot = 0,
    column_names_gp = gpar(fontsize = 20),
    column_names_centered = T,
    row_names_gp = gpar(fontsize = 15),
    # heatmap_legend_param = list(
    #     title = "log10(expr)",
    #     labels_gp = gpar(fontsize = labels_size),
    #     title_gp = gpar(fontsize = title_size)
    # ),
    column_title = "IgG",
    column_title_gp = gpar(fontsize = 30)
)
dev.off()
```


```{r fig.height=6, fig.width=8}
temp = data1[data1$Name %in% gene_drug_luad & data1$source == "SNR635", ]

temp = reshape2::dcast(temp, Name~variable, value.var = "value", fun.aggregate = mean)
rownames(temp) <- temp$Name
temp <- temp[, colnames(temp) != "Name"]

temp <- t(scale(t(temp)))
col_fun <- colorRamp2(c(-1, 0, 1), c("blue", "white", "red"))


pdf(here("gene_drug_vln/heatmap_SNR635_IgM.pdf"), width = 6, height = 6) 
Heatmap(
    temp,
    col = col_fun,
    name = "Expr",
    column_names_rot = 0,
    column_names_gp = gpar(fontsize = 20),
    column_names_centered = T,
    row_names_gp = gpar(fontsize = 15),
    # heatmap_legend_param = list(
    #     title = "log10(expr)",
    #     labels_gp = gpar(fontsize = labels_size),
    #     title_gp = gpar(fontsize = title_size)
    # ),
    column_title = "IgM",
    column_title_gp = gpar(fontsize = 30)
)
dev.off()
```


First round of protein
```{r fig.height=6, fig.width=6}

## SNR532

for(i in intersect(data1$Name, gene_drug_luad)) {
    print(i)
    
    p <- ggviolin(
        data1[data1$Name == i & data1$source == "SNR532", ], 
        x="variable", 
        y="value", 
        fill = "variable",
        add = "boxplot",
        palette = c("#00AFBB", "#E7B800", "#FC4E07"),
        add.params = list(fill = "white"),
        title = paste0(i, "(IgG)")
    ) +
        stat_compare_means(
            comparisons = list(c("LUAD", "Normal"), c("LUSC", "Normal"), c("LUAD", "LUSC")),
            method = "wilcox.test"
        ) +
        scale_y_log10()
    
    print(p)   
}


```


```{r fig.height=18, fig.width=18}

## SNR635

ggviolin(
    data1[data1$Name %in% gene_drug_luad & data1$source == "SNR635", ], 
    x="variable", 
    y="value", 
    fill = "variable",
    add = "boxplot",
    palette = c("#00AFBB", "#E7B800", "#FC4E07"),
    add.params = list(fill = "white")
) +
    stat_compare_means(
        comparisons = list(c("LUAD", "Normal"), c("LUSC", "Normal")),
        method = "t.test"
    ) +
    scale_y_log10() +
    facet_wrap(.~Name, scales = "free")
```

Second round of protein
```{r fig.height=18, fig.width=18}

## SNR532

ggviolin(
    data2[data2$Name %in% gene_drug_luad & data1$source == "SNR532", ], 
    x="variable", 
    y="value", 
    fill = "variable",
    add = "boxplot",
    palette = c("#00AFBB", "#E7B800", "#FC4E07"),
    add.params = list(fill = "white")
) +
    stat_compare_means(
        comparisons = list(c("LUAD", "Normal"), c("LUSC", "Normal")),
        method = "t.test"
    ) +
    scale_y_log10() +
    facet_wrap(.~Name, scales = "free")
```

```{r fig.height=18, fig.width=18}

## SNR635

ggviolin(
    data2[data2$Name %in% gene_drug_luad & data1$source == "SNR635", ], 
    x="variable", 
    y="value", 
    fill = "variable",
    add = "boxplot",
    palette = c("#00AFBB", "#E7B800", "#FC4E07"),
    add.params = list(fill = "white")
) +
    stat_compare_means(
        comparisons = list(c("LUAD", "Normal"), c("LUSC", "Normal")),
        method = "t.test"
    ) +
    scale_y_log10() +
    facet_wrap(.~Name, scales = "free")
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.