---
title: "TCGA"
author: "Yiming Zhang"
date: "2020/3/4"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

#root.dir = "LungCancer10x/TCGA/"
root.dir = "/mnt/raid61/Personal_data/chenli/myfile/zym/TCGA/"
full.path <- function(...) { return(paste(root.dir, ..., sep = "/")) }
```


```{r cars, include=FALSE}
library(TCGAbiolinks)
library(clusterProfiler)
library(org.Hs.eg.db)
library(survival)
library(GenomicRanges)
library(gaia)
library(openxlsx)
library(dplyr)
library(stringr)
library(progress)
```


## Download data
```{R eval=FALSE, include=FALSE}
## Prepare data
query <- GDCquery(
    project = "TCGA-LUAD",
    legacy = FALSE, 
    experimental.strategy = "RNA-Seq", 
    data.category = "Transcriptome Profiling", 
    data.type = "Gene Expression Quantification", 
    workflow.type = "HTSeq - Counts"
)

# Sys.setenv(https_proxy="http://127.0.0.1:1086")
# Sys.setenv(http_proxy="http://127.0.0.1:1086")
GDCdownload(query)

luad <- GDCprepare(query)

clinical <- GDCquery_clinic(project = "TCGA-LUAD", type = "clinical")

saveRDS(luad, "LUAD_counts.rds")
saveRDS(clinical, "LUAD_clinical.rds")


query <- GDCquery(
    project = "TCGA-LUSC",
    legacy = FALSE, 
    experimental.strategy = "RNA-Seq", 
    data.category = "Transcriptome Profiling", 
    data.type = "Gene Expression Quantification", 
    workflow.type = "HTSeq - Counts"
)

GDCdownload(query)

lusc <- GDCprepare(query)

clinical <- GDCquery_clinic(project = "TCGA-LUSC", type = "clinical")
saveRDS(lusc, "LUSC_counts.rds")
saveRDS(clinical, "LUSC_clinical.rds")


### CNV
query <- GDCquery(
    project = c("TCGA-LUAD", "TCGA-LUSC"),
    data.category = "Copy Number Variation",
    legacy = F,
    file.type = "nocnv_grch38.seg.v2.txt",
    sample.type = c("Primary solid Tumor"),
)
GDCdownload(query)
lung.no.cnv = GDCprepare(query)
saveRDS(lung.no.cnv, "Lungnocnvhg38.rds")



### DNA methgylation
query <- GDCquery(
    project = c("TCGA-LUAD","TCGA-LUSC"),
    data.category = "DNA Methylation",
    legacy = F
)

GDCdownload(query)
met <- GDCprepare(query, save = FALSE)
saveRDS(met, "LungMet.rds")

### 
# Get gene information from GENCODE using biomart
genes <- TCGAbiolinks:::get.GRCh.bioMart(genome = "hg38") 
saveRDS(genes, "hg38_genes.rds")
```

## Including Plots

You can also embed plots, for example:

```{r pressure, echo=FALSE}
## Surv
read_expr <- function(path) {
    library(SummarizedExperiment)
    library(stringr)
    library(clusterProfiler)
    library(org.Hs.eg.db)
    
    expr <- readRDS(path)
    
    expr <- as.data.frame(assay(expr))
    
    # rownames(expr) <- make.unique(sapply(rownames(expr), function(x) {
    #     str_split(x, "\\|")[[1]][1]
    # }))

    et <- bitr(rownames(expr), fromType = "ENSEMBL", toType = "SYMBOL", OrgDb = org.Hs.eg.db)
    expr <- expr[et$ENSEMBL, ]
    rownames(expr) <- make.unique(et$SYMBOL)
    
    colnames(expr) <- sapply(colnames(expr), function(x) {
        temp = str_split(x, "-")[[1]]
        
        paste(temp[1:3], collapse = "-")
    })
    
    expr
}


make_surv_plot <- function(clinical, title, subtitle="") {
    p <- TCGAanalyze_survival(
        clinical,
        "group",
        filename = NULL,
        risk.table = F,
        conf.int = F,
        ncensor.plot = TRUE,
        labels = c("high", "low"),
        color = c("red", "blue"),
        conf.int.style = "step",  # customize style of confidence intervals,
        pval.coord = c(0.1, 0.1)
    )
    p <- p$plot
    p <- p + 
        labs(title = title, subtitle = subtitle) +
        ggplot2::theme(
            aspect.ratio = 1,
            axis.text = element_text(size = 15),
            axis.title = element_text(size = 15),
            legend.position = c(0.75, 0.9),
            legend.title = element_blank(),
            legend.text = element_text(size = 15),
            plot.title = element_text(size = 15, hjust = 0.5)
        )
    
    return(p)
}


surv_plot_by_gene <- function(clinical, expr, gene, by = 3, title="") {
    library(TCGAbiolinks)
    library(ggplot2)
    
    temp = as.numeric(as.character(expr[gene, ]))
    
    if (by == "mean") {
        high <- colnames(expr)[expr[gene, ] > mean(temp)]
    } else if (by > 1) {
        high <- colnames(expr)[expr[gene, ] > summary(temp)[by]]
    } else {
        high <- expr[gene, ]
        high <- sort(high, decreasing = T)
        high <- names(high)[1:round(length(high) * by)]
    }
    
    clinical$group = "low"
    clinical[high, "group"] = "high"
    clinical$status = sapply(clinical$vital_status, function(x) { ifelse(x == "Alive", 0, 1) })
    
    return(make_surv_plot(clinical, title))
}


surv_plot_by_gene_set <- function(clinical, expr, gene_set, by = 3, title = "", by.mean = T) {
    library(TCGAbiolinks)
    library(ggplot2)
    
    if (length(setdiff(gene_set, rownames(expr))) > 0) {
        print(setdiff(gene_set, rownames(expr)))
    }

    gene_set <- intersect(gene_set, rownames(expr))
    
    if (by < 1) {
        expr = t(expr[gene_set, ])
        
        expr <- expr[order(apply(expr, 1, mean), decreasing = T), ]
        
        high <- rownames(expr)[1:round(nrow(expr) * by)]
    } else {
        
        if (by.mean) {
            expr = as.data.frame(t(expr[gene_set, ]))
            expr$mean <- apply(expr, 1, mean)
    
            high <- rownames(expr)[expr$mean > summary(expr$mean)[by]]
            high <- sapply(high, function(x) {
                str_replace_all(x, "\\.", "-")
            })
        } else {
            high = rep(TRUE, ncol(expr))
            
            for (i in gene_set) {
                temp = as.numeric(as.character(expr[i, ]))
                high <- high & expr[i, ] > summary(temp)[by]
            }
            
            high <- colnames(expr)[high]
        }

    }
    
    clinical$group = "low"
    clinical[high, "group"] = "high"
    
    return(make_surv_plot(clinical, title))
}

```


```{R}
luad_cli <- readRDS(full.path("LUAD_clinical.rds"))
rownames(luad_cli) <- luad_cli$submitter_id
luad_expr <- read_expr(full.path("LUAD_counts.rds"))
luad_fpkm <- read_expr(full.path("LUAD.rds"))

lusc_cli <- readRDS(full.path("LUSC_clinical.rds"))
rownames(lusc_cli) <- lusc_cli$submitter_id
lusc_expr <- read_expr(full.path("LUSC_counts.rds"))
lusc_fpkm <- read_expr(full.path("LUSC.rds"))
```


### Marker genes
```{R eval=FALSE, include=FALSE}
p <- surv_plot_by_gene(luad_cli, luad_fpkm, "BMP3", by = 2, title = "BMP3 (LUAD)")
ggsave(filename = full.path("BMP3_LUAD.pdf"),plot = p, width = 4, height = 4)

p <- surv_plot_by_gene(luad_cli, luad_expr, "PIGR", by = 4, title = "PIGR (LUAD)")
ggsave(filename = full.path("PIGR_LUAD.pdf"), plot = p, width = 4, height = 4)


p <- surv_plot_by_gene(luad_cli, luad_fpkm, "SCGB3A2", by = 3, title = "SCGB3A2 (LUAD)")
ggsave(filename = full.path("SCGB3A2_LUAD.pdf"), plot = p, width = 4, height = 4)


p <- surv_plot_by_gene(luad_cli, luad_expr, "CYP2B7P", by = 4, title = "CYP2B7P (LUAD)")
ggsave(filename = full.path("CYP2B7P_LUAD.pdf"), plot = p, width = 4, height = 4)

p <- surv_plot_by_gene(lusc_cli, lusc_fpkm, "IGFBP7", by = 3, title = "IGFBP7 (LUSC)")
ggsave(filename = full.path("IGFBP7_LUSC.pdf"), plot = p, width = 4, height = 4)

p <- surv_plot_by_gene(lusc_cli, lusc_expr, "VIM", by = 3, title = "VIM (LUSC)")
ggsave(filename = full.path("VIM_LUSC.pdf"), plot = p, width = 4, height = 4)
```

### gene set
```{R eval=FALSE, include=FALSE}
p <- surv_plot_by_gene_set(luad_cli, luad_expr, c("MDFI", "RPS5", "SLC40A1", "MDK", "LRRK2"), by = 5)
ggsave(filename = full.path("top5_network_LUAD_Normal.pdf"), plot = p, width = 4, height = 4)

p <- surv_plot_by_gene_set(luad_cli, luad_expr, c("NPM1", "GADPH", "MUC1", "MIF", "SUMO2"), by = 4, by.mean = F)

ggsave(filename = full.path("top5_network_LUAD.pdf"), plot = p, width = 4, height = 4)


p <- surv_plot_by_gene_set(lusc_cli, lusc_expr, c("PSMB5", "RPSA", "PSME2", "CKS1B", "C1QBP"), by = 0.8)


p <- surv_plot_by_gene_set(luad_cli, luad_fpkm, c("ZFP36L1", "HSPA1B", "RPS4Y1", "PSMA4", "RPLP0"), by = 2)
ggsave(filename = full.path("top5_network_LUAD_I.pdf"), plot = p, width = 4, height = 4)

p <- surv_plot_by_gene_set(lusc_cli, lusc_expr, c("CWC22", "ASF1A","CTBP1","SECISBP2","CBY1"), by = .8)
ggsave(filename = full.path("top5_network_LUSC_I.pdf"), plot = p, width = 4, height = 4)
```


###
```{r eval=FALSE, include=FALSE}
data = read.csv(
    full.path("../ATII_Basal/network/ATII_network_BC_clt.csv"), 
    row.names = 1, stringsAsFactors = F
)
# 2 by = 3, by.mean = T
# 18 by = 3, by.mean = F
# 19 by = 5, by.mean = F

for (i in unique(data$Clt)) {
    tryCatch({
        p <- surv_plot_by_gene_set(
            luad_cli,  luad_expr, gene_set = data$gene[data$Clt == i], 
            by = 5, title = paste0("ATII network cluster (", i, ")"), by.mean = F
        )
        
        ggsave(filename = full.path(paste0("network/ATII/", i, ".pdf")), plot = p,width = 4, height = 4)
    }, error = function(e) {
        
    })

}

i = 2
p <- surv_plot_by_gene_set(
    luad_cli,  luad_expr,  gene_set = data$gene[data$Clt == i], 
    by = 3, title = paste0("ATII network cluster (", i, ")"), by.mean = T
)

ggsave(filename = full.path(paste0("network/ATII/", i, ".pdf")), plot = p, width = 4,height = 4)

## Basal network cluster

data = read.csv(
    full.path("../ATII_Basal/network/Basal_network_BC_clt.csv"), 
    row.names = 1, stringsAsFactors = F
)
# 10 by = 2, by.mean = F 0.0063


for (i in unique(data$Clt)) {
    tryCatch({
        p <- surv_plot_by_gene_set(
            lusc_cli,  lusc_expr,  gene_set = data$gene[data$Clt == i], 
            by = 2, title = paste0("Basal network cluster (", i, ")"), by.mean = F
        )
        
        ggsave(filename = full.path(paste0("network/Basal/", i, ".pdf")), plot = p, width = 4, height = 4)
    }, error = function(e) {
        print(e)
    })
    
}
```


```{R eval=FALSE, include=FALSE}
data <- read.csv(
    full.path("../ATII_Basal/ATII_all_cells_cluster_markers.csv"), 
    row.names = 1, stringsAsFactors = F
)
data <- data %>%
    filter(p_val_adj < 0.05) %>%
    group_by(ident) %>%
    top_n(10, wt = avg_logFC) %>%
    as.data.frame()

# 2 and 4 by = 5, by.mean = T 0.0082 0.028
# 4 by = 4, by.mean = T, 0.0012
# 4 by = 3, by.mean = T, 0.04
# 2 by = .25 0.034


for (i in unique(data$ident)) {
    tryCatch({
        p <- surv_plot_by_gene_set(
            luad_cli, luad_expr, gene_set = data$gene[data$ident == i], 
            by = 5, title = paste0("ATII cluster ", i), by.mean = T
        )
        
        ggsave(filename = full.path(paste0("cluster/ATII/", i, ".pdf")), plot = p, width = 4, height = 4)
    }, error = function(e) {
        print(e)
    })
    
}

i = 3
# .65
p <- surv_plot_by_gene_set(
    luad_cli, luad_expr, gene_set = data$gene[data$ident == i], 
    by = .65, title = paste0("ATII cluster ", i), by.mean = T
)

ggsave(filename = full.path(paste0("cluster/ATII/", i, ".pdf")), plot = p, width = 4, height = 4)


i = 4
p <- surv_plot_by_gene_set(
    luad_cli, luad_expr, gene_set = data$gene[data$ident == i], 
    by = 4, title = paste0("ATII cluster ", i), by.mean = T
)
ggsave(filename = full.path(paste0("cluster/ATII/", i, ".pdf")), plot = p, width = 4, height = 4)

data <- read.csv(
    full.path("../ATII_Basal/Basal_all_cells_cluster_markers.csv"), 
    row.names = 1, stringsAsFactors = F
)
data <- data %>%
    filter(p_val_adj < 0.05) %>%
    group_by(ident) %>%
    top_n(10, wt = avg_logFC) %>%
    as.data.frame()

# 5 by = 3, by.mean = T 0.031
# 1 by = 2, by.mean = T 0.013
# 5 by = 2, by.mean = F 0.0083
# 1&5 by = 3, by.mean = F 0.034 0.046
# 5 by = 0.5 0.024
# 4 by = 0.25 0.041

for (i in unique(data$ident)) {
    tryCatch({
        p <- surv_plot_by_gene_set(
            lusc_cli, lusc_expr, gene_set = data$gene[data$ident == i], 
            by = .75, title = paste0("Basal cluster ", i), by.mean = F
        )
        
        ggsave(filename = full.path(paste0("cluster/Basal/", i, ".pdf")), plot = p, width = 4, height = 4)
    }, error = function(e) {
        print(e)
    })
    
}

i = 1
p <- surv_plot_by_gene_set(
    lusc_cli, lusc_expr, gene_set = data$gene[data$ident == i], 
    by = 2, title = paste0("Basal cluster ", i), by.mean = T
)

ggsave(filename = full.path(paste0("cluster/Basal/", i, ".pdf")), plot = p, width = 4, height = 4)


i = 4
p <- surv_plot_by_gene_set(
    lusc_cli, lusc_expr, gene_set = data$gene[data$ident == i], 
    by = .25, title = paste0("Basal cluster ", i), by.mean = T
)

ggsave(filename = full.path(paste0("cluster/Basal/", i, ".pdf")), plot = p, width = 4, height = 4)


i = 5
p <- surv_plot_by_gene_set(
    lusc_cli,  lusc_expr, gene_set = data$gene[data$ident == i], 
    by = .5, title = paste0("Basal cluster ", i), by.mean = T
)

ggsave(filename = full.path(paste0("cluster/Basal/", i, ".pdf")), plot = p, width = 4, height = 4)
```


```{R eval=FALSE, include=FALSE}
genes = c(
    "YBX2",    "DLX3" ,   "TFAP2C" , "DLX4" ,   "OTX1" ,   "SPDEF" ,  "SOX4" ,   "FEZF1",   "E2F8"   ,
    "FOXA1" ,  "MNX1"  ,  "TFAP2A" , "DMRTA2" , "HOXB13" , "PITX2" ,  "ONECUT1", "FOXE1" , "DMBX1" , 
    "PLXNB3",  "HOXA10" 
)

for(i in genes) {
    p <- surv_plot_by_gene(luad_cli, luad_expr, i, title = paste0(i, "(LUAD)"))
    print(p)
}


genes = c(
    "DLX6" ,  "DPF1" ,  "DEPDC1", "ZIC2" ,  "RAD51" , "PITX1" , "TP63" ,  "DLX5" ,  "HOXA13", "POU6F2",
    "FOXM1" , "SOX21", "LMX1B" , "ZIC5" ,  "GRHL3" , "SIM2" ,  "SIX2" ,  "OVOL1" , "DMRT2" , "FOXE1", 
    "SOX2"  , "FOXL2" , "HOXA10"
)


for(i in genes) {
    p <- surv_plot_by_gene(lusc_cli, lusc_expr, i, title = paste0(i, "(LUSC)"))
    print(p)
}
```


## CNV
```{r}
is.integer0 <- function(x) {
    is.integer(x) && length(x) == 0L
}


load_cnv <- function(segmentation_matrix, markers_list, num_of_samples){
    
    message("Loading Copy Number Data");
    
    # Detect the chromosomes
    chromosomes <- as.numeric(sort(unique(names(markers_list))));
    chromosomes <- chromosomes[which(!is.na(chromosomes))];
    
    # Detect the aberration kinds
    aberration_kinds <- 1:length(unique(segmentation_matrix[,6]));
    names(aberration_kinds) <- sort(unique(segmentation_matrix[,6]));
    
    # Detect the samples
    samples <- 1:num_of_samples;
    if(!is.numeric(segmentation_matrix[,1])){
        sample_names <- unique(segmentation_matrix[,1]);
        for(i in 1:length(sample_names)){
            segmentation_matrix[which(segmentation_matrix[,1]==sample_names[i]),1] <- i;
        }
    }
    
    region_list <- list();
    # Create the final structure list of the returned list
    
    pb <- progress_bar$new(total = length(aberration_kinds))
    for(k in 1:length(aberration_kinds)){
        pb$tick()
        region_list[[ aberration_kinds[k] ]] <- list();
        
        for(i in 1:length(chromosomes) ){
            region_list[[ aberration_kinds[k] ]][[ chromosomes[i] ]] <- matrix(0, length(samples), ncol(markers_list[[ chromosomes[i] ]]));
            rownames(region_list[[ aberration_kinds[k] ]][[ chromosomes[i] ]]) <- samples;
            colnames(region_list[[ aberration_kinds[k] ]][[ chromosomes[i] ]]) <- c(1:ncol(markers_list[[ chromosomes[i] ]]));
        }
    }
    
    pb <- progress_bar$new(total = length(aberration_kinds))
    for(k in 1:length(aberration_kinds)){
        pb$tick()
        
        ab_ids <- which(segmentation_matrix[,6]==names(aberration_kinds[k]));
        
        # In this matrix all regions aberrant as aberration_kinds[k] are stored
        tmp_matrix1 <- segmentation_matrix[ab_ids,];
        if(class(tmp_matrix1)=="numeric"){
            tmp_matrix1 <- t(as.matrix(tmp_matrix1));
        }
        for(i in 1:length(chromosomes) ){
            
            # In this matrix all regions aberrant as aberration_kinds[k] for the i-th chromsome are stored
            tmp_matrix2 <- tmp_matrix1[which(tmp_matrix1[,2]==chromosomes[i]),];
            if(class(tmp_matrix2)=="numeric"){
                tmp_matrix2 <- t(as.matrix(tmp_matrix2));
            }
            # message(".", appendLF = FALSE);
            for(j in 1:length(samples)){
                #In this matrix all regions aberrant as aberration_kinds[k] for the i-th chromsome and for the j-th sample are stored
                tmp_matrix3 <- tmp_matrix2[which(tmp_matrix2[,1]==samples[j]),];
                if(class(tmp_matrix3)=="numeric"){
                    tmp_matrix3 <- t(as.matrix(tmp_matrix3));
                }
                if(nrow(tmp_matrix3)>0){
                    for(t in 1:nrow(tmp_matrix3)){
                        start_prob <- tmp_matrix3[t,3];
                        end_prob <- tmp_matrix3[t,4];
                        
                        start_index <- which(markers_list[[ chromosomes[i] ]][1,] == start_prob);
                        end_index <- which(markers_list[[ chromosomes[i] ]][2,] == end_prob);
                        
                        if (is.integer0(start_index) || is.integer0(end_index)) {
                            next
                        }
                        
                        region_list[[ aberration_kinds[k] ]][[ chromosomes[i] ]][samples[j], start_index:end_index] <- 1;
                    }
                }
            }
        }
    }	
    message("\nDone");
    names(region_list) <- names(aberration_kinds);
    return(region_list);
    
}
```


```{r eval=FALSE, include=FALSE}
markersMatrix <-  readr::read_tsv(full.path("snp6.na35.liftoverhg38.txt"), col_names = T, col_types = "ccn", progress = FALSE)
saveRDS(markersMatrix, file = full.path("markersMatrix.rds"))


commonCNV <- readr::read_tsv(full.path("CNV.hg38.bypos.111213.bed"), progress = FALSE)
commonCNV$Chromosome <- sapply(commonCNV$Chromosome, function(x) {
    x = str_replace_all(x, "^chr", "")
    if (x == "X") {
        return(23)
    } else if (x == "Y") {
        return(24)
    } else {
        return(as.numeric(x))
    }
})

saveRDS(commonCNV, full.path("commonCNV.rds"))
```

```{r}
lung.no.cnv = readRDS("Lungnocnvhg38.rds")
markersMatrix <- readRDS("markersMatrix.rds")
commonCNV <- readRDS("commonCNV.rds")

tree <- read.csv("All_tree.csv", row.names = 1, stringsAsFactors = F)
tree <- tree[tree$Source == "TCGA", ]
tree$ID <- sapply(rownames(tree), function(x) {
    x = str_split(x, "\\.")[[1]][1:3]
    paste(x, collapse = "-")
})

# Add label (0 for loss, 1 for gain)
cnvMatrix <- cbind(lung.no.cnv,Label=NA)

cnvMatrix[cnvMatrix[,"Segment_Mean"] < -0.3,"Label"] <- 0
cnvMatrix[cnvMatrix[,"Segment_Mean"] > 0.3,"Label"] <- 1
cnvMatrix <- cnvMatrix[!is.na(cnvMatrix$Label),]

# Remove "Segment_Mean" and change col.names
cnvMatrix <- cnvMatrix[,-6]
colnames(cnvMatrix) <- c("Sample.Name", "Chromosome", "Start", "End", "Num.of.Markers", "Sample", "Aberration")

# Substitute Chromosomes "X" and "Y" with "23" and "24"
cnvMatrix[cnvMatrix$Chromosome == "X","Chromosome"] <- 23
cnvMatrix[cnvMatrix$Chromosome == "Y","Chromosome"] <- 24
cnvMatrix$Chromosome <- as.integer(cnvMatrix$Chromosome)

# Recurrent CNV identification with GAIA
colnames(markersMatrix) <- c("Probe.Name", "Chromosome", "Start")
# unique(markersMatrix$Chromosome)


markersMatrix[markersMatrix$Chromosome == "X","Chromosome"] <- 23
markersMatrix[markersMatrix$Chromosome == "Y","Chromosome"] <- 24
markersMatrix$Chromosome <- as.integer(markersMatrix$Chromosome)
markerID <- paste(markersMatrix$Chromosome,markersMatrix$Start, sep = ":")

# Removed duplicates
markersMatrix <- markersMatrix[!duplicated(markerID),]
# Filter markersMatrix for common CNV
markerID <- paste(markersMatrix$Chromosome,markersMatrix$Start, sep = ":")


commonID <- paste(commonCNV$Chromosome,commonCNV$Start, sep = ":")
markersMatrix_fil <- markersMatrix[!markerID %in% commonID,]
#cnvMatrix <- cnvMatrix[commonID %in% markerID, ]
#cnvMatrix <- as.data.frame(as.matrix(cnvMatrix))
cnvMatrix$ID <- sapply(cnvMatrix$Sample, function(x) {
    x = str_split(x, "-")[[1]][1:3]
    paste(x, collapse = "-")
})
```

```{r}
set.seed(200)
markers_obj <- load_markers(as.data.frame(markersMatrix_fil))

cnv_objs = list()
gaia_res = list()
for(i in unique(tree$Disease)) {
    for (j in unique(tree$Clt)) {
        temp_tree = tree[tree$Clt == j & tree$Disease == i, ]
        if(nrow(temp_tree) >0){
        temp_cnvMatrix <- cnvMatrix[cnvMatrix$ID %in% temp_tree$ID, ]
        
        nbsamples <- unique(temp_cnvMatrix$Sample.Name)
        
        #if (length(nbsamples) > 100) {
        #    nbsamples = nbsamples[sample(1:length(nbsamples), 100)]
        #}

        print(paste(j, i, length(nbsamples)))
        
        temp_cnvMatrix = temp_cnvMatrix[
            temp_cnvMatrix$Sample.Name %in% nbsamples,
            c("Sample.Name", "Chromosome", "Start", "End", "Num.of.Markers", "Aberration")
        ]
        
        for (k in colnames(temp_cnvMatrix)) {
            if (k %in% c("Sample.Name", "Aberration")) {
                temp_cnvMatrix[, k] <- as.character(temp_cnvMatrix[, k])
            } else {
                temp_cnvMatrix[, k] <- as.numeric(as.character(temp_cnvMatrix[, k]))
            }
        }

        cnv_obj <- load_cnv(temp_cnvMatrix, markers_obj, length(nbsamples))
        cnv_objs[[paste(i, j, sep = "_")]] <- cnv_obj
        
       
        results <- runGAIA(
            cnv_obj,
            markers_obj,
            output_file_name = full.path(paste0("GAIA_", i, "_", j, ".txt")),
            aberrations = -1,  # -1 to all aberrations
            chromosomes = -1, # -1 to all chromosomes
            approximation = TRUE, # Set to TRUE to speed up the time requirements
            num_iterations = 5000, # Reduced to speed up the time requirements
            threshold = 0.25
        )
        
         gaia_res[[paste(i, j, sep = "_")]] <- results
    }
}}

saveRDS(cnv_objs, full.path("CNV_objs_all2.rds"))
saveRDS(gaia_res, full.path("GAIA_res_all2.rds"))
```


```{r}
# Set q-value threshold
# Use a smalled value for your analysis. We set this as high values
# due to the small number of samples which did not reproduced
# results with smaller q-values
threshold <- 0.3

# Plot the results
RecCNV_list<- lapply(gaia_res,function(x){
  RecCNV <- t(apply(x,1,as.numeric))
  colnames(RecCNV) <- colnames(x)
  RecCNV <- cbind(RecCNV, score = 0)
  minval <- format(min(RecCNV[RecCNV[,"q-value"] != 0,"q-value"]), scientific = FALSE)
  minval <- substring(minval,1, nchar(minval) - 1)
  RecCNV[RecCNV[,"q-value"] == 0,"q-value"] <- as.numeric(minval)
  RecCNV[,"score"] <- sapply(RecCNV[,"q-value"],function(x)   -log10(as.numeric(x)))
  RecCNV[RecCNV[,"q-value"] == as.numeric(minval),]
  return(RecCNV)
})


pdf("LUSC_4_CNVplot_all.pdf",width = 12,height = 8)
gaiaCNVplot(RecCNV,threshold)
dev.off()

```


```{r}
# Gene annotation of recurrent CNV
library(GenomicRanges)
# Get gene information from GENCODE using biomart
genes <- TCGAbiolinks:::get.GRCh.bioMart(genome = "hg38") 
genes <- genes[genes$chromosome_name %in% c(1:22,"X","Y"),]
genes[genes$chromosome_name == "X", "chromosome_name"] <- 23
genes[genes$chromosome_name == "Y", "chromosome_name"] <- 24
genes$chromosome_name <- sapply(genes$chromosome_name,as.integer)
genes <- genes[order(genes$start_position),]
genes <- genes[order(genes$chromosome_name),]
genes <- genes[,c("external_gene_name", "chromosome_name", "start_position","end_position")]
colnames(genes) <- c("GeneSymbol","Chr","Start","End")
genes_GR <- makeGRangesFromDataFrame(genes,keep.extra.columns = TRUE)
save(genes_GR,genes,file = "genes_GR.rda", compress = "xz")


```

```{r}
##############################
## Recurrent CNV annotation ## 
##############################

### remove negative region size
AmpDel_genes_negout <- lapply(RecCNV_list,function(x) {
  sCNV <- subset(x,x[,5]>0)
  sCNV <- sCNV[sCNV[,"q-value"] <= threshold,c(1:4,6)]
  sCNV <- sCNV[order(sCNV[,3]),]
  sCNV <- sCNV[order(sCNV[,1]),]
  colnames(sCNV) <- c("Chr","Aberration","Start","End","q-value")
  sCNV_GR <- makeGRangesFromDataFrame(sCNV,keep.extra.columns = TRUE)
  
  #hits <- findOverlaps(sCNV_GR, genes_GR, type = "within")
  #sCNV_ann <- cbind(sCNV[queryHits(hits),],genes[subjectHits(hits),])
  hits <- findOverlaps(genes_GR, sCNV_GR, type = "within")
  sCNV_ann <- cbind(sCNV[subjectHits(hits),],genes[queryHits(hits),])
  AberrantRegion <- paste0(sCNV_ann[,1],":",sCNV_ann[,3],"-",sCNV_ann[,4])
  GeneRegion <- paste0(sCNV_ann[,7],":",sCNV_ann[,8],"-",sCNV_ann[,9])
  AmpDel_genes <- cbind(sCNV_ann[,c(6,2,5)],AberrantRegion,GeneRegion)
  AmpDel_genes[AmpDel_genes[,2] == 0,2] <- "Del"
  AmpDel_genes[AmpDel_genes[,2] == 1,2] <- "Amp"
  rownames(AmpDel_genes) <- NULL
  return(AmpDel_genes)
  })



### convert start to end and end to start
AmpDel_genes_trs <- lapply(RecCNV_list,function(x) {
  sCNV <- subset(x,x[,5]<0)
  colnames(sCNV)[3:4]<- colnames(sCNV)[c(4,3)]
  #sCNV[,5] <- -(sCNV[,5])
  sCNV <- sCNV[,c(1,2,4,3,5,6,7)]
  sCNV <- rbind(subset(x,x[,5]>0),sCNV)
  
  sCNV <- sCNV[sCNV[,"q-value"] <= threshold,c(1:4,6)]
  sCNV <- sCNV[order(sCNV[,3]),]
  sCNV <- sCNV[order(sCNV[,1]),]
  colnames(sCNV) <- c("Chr","Aberration","Start","End","q-value")
  sCNV_GR <- makeGRangesFromDataFrame(sCNV,keep.extra.columns = TRUE)
  
  hits <- findOverlaps(genes_GR, sCNV_GR, type = "within")
  sCNV_ann <- cbind(sCNV[subjectHits(hits),],genes[queryHits(hits),])
  AberrantRegion <- paste0(sCNV_ann[,1],":",sCNV_ann[,3],"-",sCNV_ann[,4])
  GeneRegion <- paste0(sCNV_ann[,7],":",sCNV_ann[,8],"-",sCNV_ann[,9])
  AmpDel_genes <- cbind(sCNV_ann[,c(6,2,5)],AberrantRegion,GeneRegion)
  AmpDel_genes[AmpDel_genes[,2] == 0,2] <- "Del"
  AmpDel_genes[AmpDel_genes[,2] == 1,2] <- "Amp"
  rownames(AmpDel_genes) <- NULL
  return(AmpDel_genes)
})

unlist(lapply(AmpDel_genes_trs,function(x){nrow(x)}))
save(AmpDel_genes_negout, AmpDel_genes_trs, file = "AmpDel_genes.rda")

library(openxlsx)
write.xlsx(AmpDel_genes_negout,file='AmpDel_genes_removenegative.xlsx',
  colNames = TRUE,colWidths = rep("auto",8))
write.xlsx(AmpDel_genes_trs,file='AmpDel_genes_convert.xlsx',
  colNames = TRUE,colWidths = rep("auto",8))


```

Visualizing multiple genomic alteration events
```{r}
library(maftools)
# Download Mutation Annotation Format (MAF) files
LUADmut <- GDCquery_Maf(tumor = "LUAD", pipelines = "mutect2")
LUSCmut <- GDCquery_Maf(tumor = "LUSC", pipelines = "mutect2")

mut <- plyr::rbind.fill(LUADmut, LUSCmut)

save(LUADmut,file ="mafMutect2LUAD.rda", compress = "xz")
save(LUSCmut,file ="mafMutect2LUSC.rda", compress = "xz")

mut$ID <- sapply(mut$Tumor_Sample_Barcode, function(x) {
    x = str_split(x, "-")[[1]][1:3]
    paste(x, collapse = "-")
})
```

```{r}
# To prepare for maftools we will also include clinical data
# For a mutant vs WT survival analysis 
# get indexed clinical patient data for LUAD samples
LUAD_clin <- GDCquery_clinic(project = "TCGA-LUAD", type = "Clinical")
# LUAD_clin <- readRDS("LUAD_clinical.rds")

# get indexed clinical patient data for LUSC samples
LUSC_clin <- GDCquery_clinic(project = "TCGA-LUSC", type = "Clinical")
# LUSC_clin <- readRDS("LUSC_clinical.rds")

# Bind the results, as the columns might not be the same,
# we will will plyr rbind.fill, to have all columns from both files
clinical <- plyr::rbind.fill(LUAD_clin,LUSC_clin)
colnames(clinical)[1] <- "Tumor_Sample_Barcode"
clinical$Overall_Survival_Status <- 1
clinical$Overall_Survival_Status[which(clinical$vital_status != "dead")] <- 0
clinical$time <- clinical$days_to_death
clinical$time[is.na(clinical$days_to_death)] <- clinical$days_to_last_follow_up[is.na(clinical$days_to_death)]

# Create object to use in maftools

maf_res <- list()
for(i in unique(tree$Disease)) {
    for (j in unique(tree$Clt)) {
        temp_tree = tree[tree$Clt == j & tree$Disease == i, ]
        if(nrow(temp_tree) >0){
          print(paste(j, i, length(temp_tree$ID)))
          sub_mut <- mut[mut$ID %in% temp_tree$ID, ]
          sub_clinical <- clinical[clinical$Tumor_Sample_Barcode %in% temp_tree$ID,]
          maf <- read.maf(maf = sub_mut, clinicalData = sub_clinical, isTCGA = T)
          
          maf_res[[paste(i, j, sep = "_")]] <- maf
        }
    }
  }

save(maf_res,file="maf_res.rda")
```

```{r}
pdf("mafSummaryPlot.pdf",width=9,height =5)
for (i in maf_res){
  plotmafSummary(maf = i,
               rmOutlier = TRUE,
               addStat = 'median',
               dashboard = TRUE)
}
dev.off()

maf_gene_summary <- lapply(maf_res,function(x){
  tmp <- x@gene.summary
  tmp$Totalsample <- as.numeric(x@summary$summary[3])
  tmp$MutatedSamples <- as.numeric(tmp$MutatedSamples)
  tmp$MutPercentage <- tmp$MutatedSamples/tmp$Totalsample
  tmp <- tmp[order(tmp$total,decreasing = T),]
  return(tmp)
  })

write.xlsx(maf_gene_summary,file='maf_gene_summary.xlsx',
  colNames = TRUE,colWidths = rep("auto",8))

```


```{r}
## survival analysis
plot <- mafSurvival(maf = maf_res[[1]],
                   genes = "EGFR",
                   fn = NULL,
                   time = 'time',
                   Status = 'Overall_Survival_Status',
                   isTCGA = TRUE)


```

## CircosPlot
```{r}
###############################################
## Genomic aberration overview - Circos plot ## 
###############################################

TF_cnv <- lapply(2:5,function(x){ read.xlsx("tf_with_cnv.xlsx",sheet = x)})

TF_cnv_fil <- lapply(TF_cnv,function(x){
  subset(x, x$Aberration=="Amp" & x$'q-value'<0.05)
})
names(TF_cnv_fil) <- c("LUAD_3_remove","LUAD_3_convert","LUSC_3_remove","LUSC_3_convert")

# Load recurrent CNV data for selected cancer 
load("CNV_results.rda")

```

```{r}
##  LUAD_3 remove
s.cnv_res <- lapply(TF_cnv_fil[c(1,2)],function(x){
  filt_ID <- x$AberrantRegion
  CNV_ID <- paste0(RecCNV_list[[1]][,1],":",RecCNV_list[[1]][,3],"-",RecCNV_list[[1]][,4])
  s.cnv <- RecCNV_list[[1]][CNV_ID %in% filt_ID,]

# s.cnv <- subset(RecCNV_list[[1]],RecCNV_list[[1]][,5]>0)
# Prepare selected sample CNV data for circos plot
  s.cnv <- as.data.frame(s.cnv[s.cnv[,"q-value"] <= threshold,c(1:4,6)])
  s.cnv <- s.cnv[,c(1,3,4,2)]
  s.cnv[s.cnv$Chromosome == 23,"Chromosome"] <- "X"
  s.cnv[s.cnv$Chromosome == 24,"Chromosome"] <- "Y" 
  Chromosome <- paste0("chr",s.cnv[,1])
  s.cnv <- cbind(Chromosome, s.cnv[,-1])
  s.cnv[,1] <- as.character(s.cnv[,1])
  s.cnv[,4] <- as.character(s.cnv[,4])
  s.cnv <- cbind(s.cnv,CNV=1)
  colnames(s.cnv) <- c("Chromosome","Start_position","End_position","Aberration_Kind","CNV")
  return(s.cnv)
})    

s.cnv_res[c(3,4)] <- lapply(TF_cnv_fil[c(3,4)],function(x){
  filt_ID <- x$AberrantRegion
  CNV_ID <- paste0(RecCNV_list[[5]][,1],":",RecCNV_list[[5]][,3],"-",RecCNV_list[[5]][,4])
  s.cnv <- RecCNV_list[[5]][CNV_ID %in% filt_ID,]

# s.cnv <- subset(RecCNV_list[[1]],RecCNV_list[[1]][,5]>0)
# Prepare selected sample CNV data for circos plot
  s.cnv <- as.data.frame(s.cnv[s.cnv[,"q-value"] <= threshold,c(1:4,6)])
  s.cnv <- s.cnv[,c(1,3,4,2)]
  s.cnv[s.cnv$Chromosome == 23,"Chromosome"] <- "X"
  s.cnv[s.cnv$Chromosome == 24,"Chromosome"] <- "Y" 
  Chromosome <- paste0("chr",s.cnv[,1])
  s.cnv <- cbind(Chromosome, s.cnv[,-1])
  s.cnv[,1] <- as.character(s.cnv[,1])
  s.cnv[,4] <- as.character(s.cnv[,4])
  s.cnv <- cbind(s.cnv,CNV=1)
  colnames(s.cnv) <-c("Chromosome","Start_position","End_position","Aberration_Kind","CNV")
  return(s.cnv)
})    
```

```{r}
## 4 types
TF_cnv_uniq <- lapply(TF_cnv_fil,function(x){
  TF_uniq <- subset(x,!duplicated(x$TF))
  TF_uniq <- TF_uniq[,c(1,18)]
  TF_uniq$Chromosome <- paste0("chr",do.call(rbind, strsplit(TF_uniq$GeneRegion,split="[:-]"))[,1])
  TF_uniq$Start_Position <- as.numeric(do.call(rbind, strsplit(TF_uniq$GeneRegion,split="[:-]"))[,2])
  TF_uniq$End_Position <- as.numeric(do.call(rbind, strsplit(TF_uniq$GeneRegion,split="[:-]"))[,3])
  TF_uniq$Variant_Classification <- "Transcription Factor"
  TF_uniq$Type <- 1
  TF_uniq <- TF_uniq[,c(3,4,5,7,6,1)]
  })


```

```{r}
# pdf("LUAD_3_remove_circosPlot_chr1_TF.pdf",width = 5,height = 5)
par(mar=c(1,1,1,1),cex=1.5)
circos.par("start.degree" = 90, canvas.xlim = c(0, 1), canvas.ylim = c(0, 1), gap.degree = 270, cell.padding = c(0, 0, 0, 0), track.margin = c(0.005, 0.005))
circos.initializeWithIdeogram(chromosome.index = "chr1") 
circos.par(cell.padding = c(0, 0, 0, 0))
# Add CNV results
colors <- c("forestgreen","firebrick")
names(colors)  <- c(0,1)
circos.genomicTrackPlotRegion(s.cnv_res[[1]],  ylim = c(0,1.2),
                              panel.fun = function(region, value, ...) {
                                circos.genomicRect(region, value, ytop.column = 2, ybottom = 0,
                                                   col = colors[value[[1]]], 
                                                   border="white")
                                cell.xlim = get.cell.meta.data("cell.xlim")
                                circos.lines(cell.xlim, c(0, 0), lty = 2, col = "#00000040")
                              })

# Add mutation results representing single genes

colors <- "blue"
names(colors)  <- "Transcription Factor"
circos.genomicTrackPlotRegion(TF_cnv_uniq[[1]], ylim = c(0.3,2.2), track.height = 0.05,
                              panel.fun = function(region, value, ...) {
                                circos.genomicPoints(region, value, cex = 0.4, pch = 16, col = colors[value[[2]]], ...)
                              })

circos.genomicTrackPlotRegion(TF_cnv_uniq[[1]], ylim = c(0, 1), track.height = 0.1, bg.border = NA)
i_track = get.cell.meta.data("track.index")

circos.genomicTrackPlotRegion(TF_cnv_uniq[[1]], ylim = c(0,1),
                              panel.fun = function(region, value, ...) {
                                circos.genomicText(region, value, 
                                                   y = 1, 
                                                   labels.column = 3,
                                                   col = colors[value[[2]]],
                                                   facing = "clockwise", adj = c(1, 0.5),
                                                   posTransform = posTransform.text, cex = 0.4, niceFacing = TRUE)
                              }, track.height = 0.1, bg.border = NA)

circos.genomicPosTransformLines(TF_cnv_uniq[[1]],
                                posTransform = function(region, value)
                                  posTransform.text(region, 
                                                    y = 1, 
                                                    labels = value[[3]],
                                                    cex = 0.4, track.index = i_track+1),
                                direction = "inside", track.index = i_track)

circos.clear()

legend(0.25, 0.2, bty="n", y.intersp=1, c("Amp","Del"), pch=15, 
       col=c("firebrick","forestgreen"), title="CNVs", text.font=1, cex=0.4, title.adj=0)
legend(0, 0.15, bty="n", y.intersp=1, names(colors), pch=16, 
       col=colors, text.font=1, cex=0.4, title.adj=0)

# dev.off()


```

```{r}
pdf("LUAD_3_remove_circosPlot_TF.pdf",width = 7,height = 7)
par(mar=c(1,1,1,1),cex=1.5)
#circos.par("start.degree" = 90, canvas.xlim = c(0, 1), canvas.ylim = c(0, 1), gap.degree = 270, cell.padding = c(0, 0, 0, 0), track.margin = c(0.005, 0.005))
circos.initializeWithIdeogram(chromosome.index = c("chr1","chr7","chr8","chr12","chr17","chr19","chr20")) 
circos.par(cell.padding = c(0, 0, 0, 0))
# Add CNV results
colors <- c("forestgreen","firebrick")
names(colors)  <- c(0,1)
circos.genomicTrackPlotRegion(s.cnv_res[[1]],  ylim = c(0,1.2),
                              panel.fun = function(region, value, ...) {
                                circos.genomicRect(region, value, ytop.column = 2, ybottom = 0,
                                                   col = colors[value[[1]]], 
                                                   border="white")
                                cell.xlim = get.cell.meta.data("cell.xlim")
                                circos.lines(cell.xlim, c(0, 0), lty = 2, col = "#00000040")
                              })

# Add TF genes

colors <- "blue"
names(colors)  <- "Transcription Factor"
circos.genomicTrackPlotRegion(TF_cnv_uniq[[1]], ylim = c(0.3,2.2), track.height = 0.05,
                              panel.fun = function(region, value, ...) {
                                circos.genomicPoints(region, value, cex = 0.4, pch = 16, col = colors[value[[2]]], ...)
                              })

circos.genomicTrackPlotRegion(TF_cnv_uniq[[1]], ylim = c(0, 1), track.height = 0.1, bg.border = NA)
i_track = get.cell.meta.data("track.index")

circos.genomicTrackPlotRegion(TF_cnv_uniq[[1]], ylim = c(0,1),
                              panel.fun = function(region, value, ...) {
                                circos.genomicText(region, value, 
                                                   y = 1, 
                                                   labels.column = 3,
                                                   col = colors[value[[2]]],
                                                   facing = "clockwise", adj = c(1, 0.5),
                                                   posTransform = posTransform.text, cex = 0.4, niceFacing = TRUE)
                              }, track.height = 0.1, bg.border = NA)

circos.genomicPosTransformLines(TF_cnv_uniq[[1]],
                                posTransform = function(region, value)
                                  posTransform.text(region, 
                                                    y = 1, 
                                                    labels = value[[3]],
                                                    cex = 0.4, track.index = i_track+1),
                                direction = "inside", track.index = i_track)

circos.clear()

legend(-0.2, 0.2, bty="n", y.intersp=1, c("Amp","Del"), pch=15, 
       col=c("firebrick","forestgreen"), title="CNVs", text.font=1, cex=0.4, title.adj=0)
legend(-0.2, 0, bty="n", y.intersp=1, names(colors), pch=16, 
       col=colors, text.font=1, cex=0.4, title.adj=0)

dev.off()


```


```{R}
format_mut_data <- function(mut) {
    # Prepare selected mutations data for circos plot
    s.mut <- mut[mut$mut.id %in% unique(mut.id[duplicated(mut.id)]),]
    s.mut <- s.mut[,c("Chromosome","Start_Position","End_Position","Variant_Classification","Hugo_Symbol")]
    s.mut$Variant_Classification = "Transcript Factor"
    s.mut$Type = as.numeric(as.factor(s.mut$Variant_Classification))
    s.mut <- unique(s.mut)
    s.mut <- s.mut[,c(1:3,6,4,5)]
    
    # Add mutation results representing single genes
    genes.mut <- paste0(s.mut$Hugo_Symbol,"-",s.mut$Type)
    s.mutt <- cbind(s.mut, genes.mut)
    n.mut <- table(genes.mut)
    idx <- !duplicated(s.mutt$genes.mut)
    s.mutt <- s.mutt[idx,]
    s.mutt <- cbind(s.mutt, num=n.mut[s.mutt$genes.mut])
    # genes.num <- paste0(s.mutt$Hugo_Symbol," (",s.mutt$num.Freq,")")
    genes.num = s.mutt$Hugo_Symbol
    s.mutt <- cbind(s.mutt[,-c(6:8)], genes.num)
    s.mutt[,6] <- as.character(s.mutt[,6])
    s.mutt$num.Freq <- NULL
    s.mutt
}


format_sheet_to_mut_data <- function(x) {
    TF_uniq <- subset(x,!duplicated(x$TF))
    TF_uniq <- TF_uniq[,c(1,18)]
    TF_uniq$Chromosome <- paste0("chr",do.call(rbind, strsplit(TF_uniq$GeneRegion,split="[:-]"))[,1])
    TF_uniq$Start_Position <- as.numeric(do.call(rbind, strsplit(TF_uniq$GeneRegion,split="[:-]"))[,2])
    TF_uniq$End_Position <- as.numeric(do.call(rbind, strsplit(TF_uniq$GeneRegion,split="[:-]"))[,3])
    TF_uniq$Variant_Classification <- "Transcription Factor"
    TF_uniq$Type <- 1
    TF_uniq <- TF_uniq[,c(3,4,5,7,6,1)]
    TF_uniq
}


format_data_to_mut <- function(x) {
    x$Chromosome <- paste0("chr",do.call(rbind, strsplit(x$GeneRegion,split="[:-]"))[,1])
    x$Start_Position <- as.numeric(do.call(rbind, strsplit(x$GeneRegion,split="[:-]"))[,2])
    x$End_Position <- as.numeric(do.call(rbind, strsplit(x$GeneRegion,split="[:-]"))[,3])
    x$Variant_Classification <- "Transcription Factor"
    x$TF = x$external_gene_name
    x$Type <- 1
    x = x[, c("Chromosome", "Start_Position", "End_Position", "Type", "Variant_Classification", "TF")]
    unique(x)
}

format_data_for_circlize <- function(data, in_house_cnv, threshold=0.3) {
    data = data[data$Aberration == "Amp", ]
    
    s.cnv = data.frame(
        Chromosome=as.character(sapply(data$AberrantRegion, function(x) { 
            paste0("chr", str_split(x, ":")[[1]][1]) 
        })),
        Start_position=as.numeric(as.character(sapply(data$AberrantRegion, function(x) { 
            as.numeric(str_split(x, "[:-]")[[1]][2]) 
        }))),
        End_position=as.numeric(as.character(
            sapply(data$AberrantRegion, function(x) { 
            as.numeric(str_split(x, "[:-]")[[1]][3]) 
        }))),
        Aberration_Kind=rep("TCGA", nrow(data)),
        CNV=rep(1, nrow(data)),
        qval=data$qvalue,
        TF=data$external_gene_name
    )
    
    print(dim(s.cnv))
    in_house_cnv = in_house_cnv[in_house_cnv$external_gene_name %in% as.character(unique(s.cnv$TF)), ]
    in_house_cnv = in_house_cnv[, c("Chromosome","Start_position","End_position","Aberration_Kind", "external_gene_name")]
    colnames(in_house_cnv)[colnames(in_house_cnv) == "external_gene_name"] = "TF"
    in_house_cnv$CNV = 0.8
    in_house_cnv$qval = 0
    in_house_cnv$Aberration_Kind = "In-house"
    
    print(dim(in_house_cnv))
    
    rbind(s.cnv, in_house_cnv)
}
```


```{R fig.height=8, fig.width=8}
make_circos_plot <- function(s.cnv, s.mut, cex=0.4, no.legend = FALSE) {
    circos.clear()
    rownames(s.mut) <- s.mut$TF
    
    # calculate -log10(qvalue)
    s.cnv$qval = -1 * log10(s.cnv$qval)
    s.cnv$qval[is.infinite(s.cnv$qval)] = max(s.cnv$qval[is.finite(s.cnv$qval)])
 
    p05 = -1 * log10(0.05) / max(s.cnv$qval)
    p01 = -1 * log10(0.01) / max(s.cnv$qval)
    
    s.cnv$qval <- s.cnv$qval / max(s.cnv$qval)
    
    s.cnv$qval = s.cnv$qval / max(s.cnv$qval)
    
    chroms = unique(s.mut$Chromosome)
    
    if(length(chroms) == 0) {
        return()
    }

    par(mar=c(1,1,1,1), cex=1.5, family = "Arial Unicode MS")
    
    if (length(chroms) == 1) {
        circos.par(
            "start.degree" = 90, canvas.xlim = c(0, 1), canvas.ylim = c(0, 1),
             gap.degree = 270, cell.padding = c(0, 0, 0, 0), track.margin = c(0.005, 0.005)
        )
    } else {
        circos.par(gap.degree = 8)
    }

    circos.initializeWithIdeogram(chromosome.index = chroms)
    circos.par(cell.padding = c(0, 0, 0, 0))
    # Add CNV results
    colors <- c("forestgreen","firebrick")
    names(colors)  <- c("In-house", "TCGA")

    circos.genomicTrackPlotRegion(
        s.cnv,  ylim = c(0, 1.2),
         panel.fun = function(region, value, ...) {
            circos.genomicRect(
                region, value, ytop.column = 2, ybottom = 0,
                col = colors[as.character(value[[1]])], border="white"
            )
            cell.xlim = get.cell.meta.data("cell.xlim")
            circos.lines(cell.xlim, c(0, 0), lty = 2, col = "#00000040")

            if (value[[3]] > 0 && value[[4]] %in% rownames(s.mut)) {
                circos.points(
                    s.mut[value[[4]], "Start_Position"], 
                    value[[3]], col = "blue", pch=18, cex=cex
                )
            }
         }
    )
    
    for (c in chroms) {
        circos.yaxis(
            at=c(p05, p01), 
            labels.cex=1,
            labels = c('*', '**'),
            labels.niceFacing = F,
            sector.index=c
        )
    }

    circos.genomicTrackPlotRegion(
        s.mut, ylim = c(0.3, 2.2), track.height = 0.05,
        panel.fun = function(region, value, ...) {
            circos.genomicPoints(region, value, cex = cex, pch = 16, col = "blue", ...)
    })
    
    
    circos.genomicTrackPlotRegion(s.mut, ylim = c(0, 1), track.height = 0.1, bg.border = NA)
    i_track = get.cell.meta.data("track.index")
    
    circos.genomicTrackPlotRegion(
        s.mut, ylim = c(0, 1),
        panel.fun = function(region, value, ...) {
            circos.genomicText(
                region, value, y = 1, labels.column = 3,
                col = "blue", facing = "clockwise",
                adj = c(0.9, 0.5),
                posTransform = posTransform.text,
                cex = cex, niceFacing = TRUE
            )
       },
       track.height = 0.1, bg.border = NA
    )
    
    circos.genomicPosTransformLines(
        s.mut,
        posTransform = function(region, value) {
            posTransform.text(
                region, y = 0.3, labels = value[[3]],
                cex = cex, track.index = i_track + 1
            )
        },
        direction = "inside", track.index = i_track
    )
    
    circos.clear()
    
    if (length(chroms) == 1) {
        x1 = 0
        y1 = 0.25
        x2 = 0
        y2 = 0.1
    } else {
        x1 = -0.15
        y1 = 0.15
        x2 = -0.2
        y2 = -0.05
    }
    
    if (!no.legend) {
        legend(
            x1, y1, bty="n", y.intersp=1, names(colors), pch=15, 
            col=colors, title="CNVs", 
            text.font=1, cex=cex, title.adj=0
        )
        
        # legend(
        #     x2, y2, bty="n", y.intersp=1, names(colors), pch=16, 
        #     col=colors, title="Mutations", text.font=1, cex=cex, title.adj=0
        # )
    }
}
```




```{r pressure, echo=FALSE}
in_house_cnv <- readRDS(full.path("TCGA/In-house_CNV.rds"))
in_house_cnv <- in_house_cnv[in_house_cnv$Aberration_Kind == "gain", ]
in_house_cnv$Aberration_Kind = "In-house"

data = readRDS(full.path("TCGA/gene_CNV.rds"))
```

```{r}
temp = data[data$external_gene_name %in% names(contrib) & str_detect(data$Source, "LUAD"), ]
```


```{r fig.height=8, fig.width=8}
s.mut <- format_data_to_mut(temp)
s.mut <- s.mut[s.mut$End_Position > s.mut$Start_Position, ]
s.cnv = format_data_for_circlize(temp, in_house_cnv[str_detect(in_house_cnv$Source, "LUAD"), ])
s.cnv <- s.cnv[s.cnv$End_position > s.cnv$Start_position, ]
# s.mut <- format_sheet_to_mut_data(temp)

s.cnv$TF = as.character(s.cnv$TF)
s.mut$TF = as.character(s.mut$TF)

# pdf(full.path("circos_plot_ATII_high.pdf"), width = 8, height = 8)
make_circos_plot(s.cnv, s.mut, cex = 0.8)
# dev.off()
```


```{r fig.height=8, fig.width=8}
genes = c(
  "CD74", "RPS14", "CCPC69", "GPX3", "TNIP1", 
  "MISP1", "CFD", "PTBP1", "MED16", "R3HPM4",
  "CSNK1D", "OGF0D3", "CD7", "CYBC1", "NARF", "FOXK2", "TBCD"
)
temp = data[data$external_gene_name %in% genes & str_detect(data$Source, "LUAD"), ]

s.mut <- format_data_to_mut(temp)
s.mut <- s.mut[s.mut$End_Position > s.mut$Start_Position, ]
s.cnv = format_data_for_circlize(temp, in_house_cnv[str_detect(in_house_cnv$Source, "LUAD"), ])
s.cnv <- s.cnv[s.cnv$End_position > s.cnv$Start_position, ]
# s.mut <- format_sheet_to_mut_data(temp)

s.cnv$TF = as.character(s.cnv$TF)
s.mut$TF = as.character(s.mut$TF)

pdf(full.path("circos_plot_ATII_high.pdf"), width = 8, height = 8)
make_circos_plot(s.cnv, s.mut, cex = 0.8)
dev.off()
```




```{r}
temp = data[data$external_gene_name %in% c("KMT2C","HMGB3","MET","SREBF1","MET","ID2","ID3","ETV1") & data$Source == "ATII-high(225 LUAD)", ]
```


```{r fig.height=8, fig.width=8}
s.mut <- format_data_to_mut(temp)
s.mut <- s.mut[s.mut$End_Position > s.mut$Start_Position, ]
s.cnv = format_data_for_circlize(temp, in_house_cnv[str_detect(in_house_cnv$Source, "LUAD"), ])
s.cnv <- s.cnv[s.cnv$End_position > s.cnv$Start_position, ]
# s.mut <- format_sheet_to_mut_data(temp)

s.cnv$TF = as.character(s.cnv$TF)
s.mut$TF = as.character(s.mut$TF)

# pdf(full.path("circos_plot_ATII_high.pdf"), width = 8, height = 8)
make_circos_plot(s.cnv, s.mut, cex = 0.8)
# dev.off()
```


```{r}
temp = data[data$external_gene_name %in% c("KLF6","PITX1","TOX4","PSMD11") & data$Source == "Basal-high(285 LUSC)", ]
```


```{r fig.height=8, fig.width=8}
s.mut <- format_data_to_mut(temp)
s.mut <- s.mut[s.mut$End_Position > s.mut$Start_Position, ]
s.cnv = format_data_for_circlize(temp, in_house_cnv[str_detect(in_house_cnv$Source, "LUSC"), ])
s.cnv <- s.cnv[s.cnv$End_position > s.cnv$Start_position, ]
# s.mut <- format_sheet_to_mut_data(temp)

# pdf(full.path("circos_plot_ATII_high.pdf"), width = 8, height = 8)
make_circos_plot(s.cnv, s.mut, cex = 0.8)
# dev.off()
```